% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/get_signalp5.R
\name{get_signalp5}
\alias{get_signalp5}
\alias{get_signalp5.default}
\alias{get_signalp5.character}
\alias{get_signalp5.data.frame}
\alias{get_signalp5.list}
\alias{get_signalp5.AAStringSet}
\title{Query SignalP 5.0 web server.}
\source{
\url{http://www.cbs.dtu.dk/services/SignalP/}
}
\usage{
get_signalp5(data, ...)

\method{get_signalp5}{character}(
  data,
  org_type = c("euk", "gram-", "gram+", "archea"),
  splitter = 2500L,
  attempts = 2,
  progress = FALSE,
  ...
)

\method{get_signalp5}{data.frame}(data, sequence, id, ...)

\method{get_signalp5}{list}(data, ...)

\method{get_signalp5}{default}(data = NULL, sequence, id, ...)

\method{get_signalp5}{AAStringSet}(data, ...)
}
\arguments{
\item{data}{A data frame with protein amino acid sequences as strings in one column and corresponding id's in another. Alternatively a path to a .fasta file with protein sequences. Alternatively a list with elements of class "SeqFastaAA" resulting from \code{\link[seqinr]{read.fasta}} call. Alternatively an `AAStringSet` object. Should be left blank if vectors are provided to sequence and id arguments.}

\item{...}{currently no additional arguments are accepted apart the ones documented bellow.}

\item{org_type}{One of "euk", "gram-", "gram+" or "archea". Default is "euk". Are the protein sequences from Eukarya, Gram-negative Bacteria, Gram-positive Bacteria or Archaea.}

\item{splitter}{An integer indicating the number of sequences to be in each .fasta file that is to be sent to the server. Default is 2500. Change only in case of a server side error. Accepted values are in range of 1 to 5000.}

\item{attempts}{Integer, number of attempts if server unresponsive, at default set to 2.}

\item{progress}{Boolean, whether to show messages of the job id for each batch. Default is FALSE.}

\item{sequence}{A vector of strings representing protein amino acid sequences, or the appropriate column name if a data.frame is supplied to data argument. If .fasta file path, or list with elements of class "SeqFastaAA" provided to data, this should be left blank.}

\item{id}{A vector of strings representing protein identifiers, or the appropriate column name if a data.frame is supplied to data argument. If .fasta file path, or list with elements of class "SeqFastaAA" provided to data, this should be left blank.}
}
\value{
if org_type is "euk" a data frame with columns:
\describe{
  \item{id}{Character, as from input}
  \item{Prediction}{Integer, The type of signal peptide predicted: Sec/SPI, Tat/SPI, Sec/SPII or Other if no signal peptide predicted}
  \item{SP.Sec.SPI}{Numeric, marginal probability that the protein contains a Sec  N-terminal signal peptide (Sec/SPI).}
  \item{Other}{Numeric, the probability that the sequence does not have any kind of signal peptid.}
  \item{CS_pos}{Character, cleavage site position.}
  \item{Pr}{Numeric, probability of the predicted cleavage site position.}
  \item{cleave.site}{Numeric,llocal amino acid sequence arround the predicted cleavage site.}
  \item{is.signalp}{Logical, did SignalP5 predict the presence of a signal peptide.}
  \item{sp.length}{Integer, length of the predicted signal peptide.}
  }
  
if org_type is one of "gram-", "gram+" or "archea" the returned data frame will have two additional columns between `SP.Sec.SPI` and `Other`:
 
\describe{
  \item{TAT.Tat.SPI}{Numeric, marginal probability that the protein contains a Tat N-terminal signal peptide (Tat/SPI).}
  \item{LIPO.Sec.SPII}{Numeric, marginal probability that the protein contains a Lipoprotein N-terminal signal peptide (Sec/SPII).}
  }
}
\description{
The SignalP 5.0 server predicts the presence of signal peptides and the location of their cleavage sites in proteins from Archaea, Gram-positive Bacteria, Gram-negative Bacteria and Eukarya.
}
\note{
This function creates temporary files in the working directory. If something goes wrong during communication with the server and progress was set to TRUE, predictions can be obtained using `file.path("http://www.cbs.dtu.dk/services/SignalP-5.0/tmp", jobid, "output_protein_type.txt")` eg `read.delim(file.path(...), header = TRUE, skip = 1)`.
}
\examples{

library(ragp)
sp5_pred <- get_signalp5(data = at_nsp[1:10,],
                         sequence,
                         Transcript.id)
sp5_pred

}
\references{
Almagro Armenteros JJ, Tsirigos KD, SÃ¸nderby CK, Petersen TN, Winther O, Brunak S, von Heijne G, Nielsen H. (2019) SignalP 5.0 improves signal peptide predictions using deep neural networks. Nature Biotechnology, 37:420-423, doi:10.1038/s41587-019-0036-z
}
\seealso{
\code{\link[ragp]{get_signalp}} \code{\link[ragp]{get_targetp}}
}
