<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Filtering hydroxyproline rich glycoproteins • ragp</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../../pkgdown.css" rel="stylesheet">
<script src="../../pkgdown.js"></script><meta property="og:title" content="Filtering hydroxyproline rich glycoproteins">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-inverse navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../../index.html">ragp</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.3.0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../../articles/pkgdown/start.html">Getting started</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../../articles/pkgdown/filter.html">Filtering hydroxyproline rich glycoproteins</a>
    </li>
    <li>
      <a href="../../articles/pkgdown/analyse.html">Analyse hydroxyproline rich glycoproteins</a>
    </li>
    <li>
      <a href="../../articles/pkgdown/PSV.html">Protein Sequence Visualization</a>
    </li>
  </ul>
</li>
<li>
  <a href="../../reference/index.html">Functions</a>
</li>
<li>
  <a href="../../news/index.html">News</a>
</li>
<li>
  <a href="https://ragp.shinyapps.io/Rapp/">ragp shiny</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/missuse/ragp/">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Filtering hydroxyproline rich glycoproteins</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/missuse/ragp/blob/master/vignettes/pkgdown/filter.Rmd"><code>vignettes/pkgdown/filter.Rmd</code></a></small>
      <div class="hidden name"><code>filter.Rmd</code></div>

    </div>

    
    
<style type="text/css">

body{ /* Normal  */
      font-size: 15px;
  }
code.r{ /* Code block */
    font-size: 14px;
}
  
</style>
<p>The filtering layer in <code>ragp</code> consists of two sequential tasks:</p>
<ul>
<li>predict the presence of secretory signals (N-sp) and filter sequences containing them.<br>
</li>
<li>predict proline hydroxylation and filter sequences containing at least several potential hydroxyprolines.</li>
</ul>
<p>The following section explains how to perform these tasks on a set of 2700 Arabidopsis protein sequences included in <code>ragp</code> as <code>at_nsp</code> data frame.</p>
<p>Required packages:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(ragp)
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(eulerr)
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(dplyr)</code></pre></div>
<div id="predicting-n-terminal-secretory-signal-sequences-n-sp" class="section level2">
<h2 class="hasAnchor">
<a href="#predicting-n-terminal-secretory-signal-sequences-n-sp" class="anchor"></a>Predicting N-terminal secretory signal sequences (N-sp)</h2>
<p>Hydroxyproline rich glycoproteins (HRGPs) are secreted proteins, therefore they are expected to contain a secretory signal sequence on the N-terminus (N-sp). <code>ragp</code> incorporates N-sp prediction by querying <a href="http://www.cbs.dtu.dk/services/SignalP/">SignalP</a>, <a href="http://www.cbs.dtu.dk/services/TargetP/">TargetP</a> <span class="citation">(Emanuelsson et al. 2000)</span> and <a href="http://phobius.sbc.su.se/">Phobius</a> <span class="citation">(Käll, Krogh, and Sonnhammer 2007)</span> in an efficient manner via the functions: <code><a href="../../reference/get_signalp.html">get_signalp()</a></code>, <code><a href="../../reference/get_targetp.html">get_targetp()</a></code> and <code><a href="../../reference/get_phobius.html">get_phobius()</a></code>.</p>
<p>To query <a href="http://www.cbs.dtu.dk/services/SignalP/">SignalP</a> predictions:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">nsp_signalp &lt;-<span class="st"> </span><span class="kw"><a href="../../reference/get_signalp.html">get_signalp</a></span>(<span class="dt">data =</span> at_nsp, <span class="co">#data frame name</span>
                           <span class="dt">sequence =</span> sequence, <span class="co">#column containing the sequences</span>
                           <span class="dt">id =</span> Transcript.id) <span class="co">#column containing protein identifiers</span></code></pre></div>
<p>The predictions for the 2700 sequences contained in <code>at_nsp</code> data frame should be available in around 1 minute. The returned object <code>nsp_signalp</code> is a data frame resembling the web servers output:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(nsp_signalp)
<span class="co">#&gt;            id  Cmax Cmax.pos  Ymax Ymax.pos  Smax Smax.pos Smean Dmean</span>
<span class="co">#&gt; 1 ATCG00660.1 0.210       30 0.162       30 0.260        2 0.146 0.154</span>
<span class="co">#&gt; 2 AT2G43600.1 0.860       23 0.894       23 0.984       13 0.930 0.913</span>
<span class="co">#&gt; 3 AT2G28410.1 0.779       23 0.826       23 0.940       15 0.877 0.853</span>
<span class="co">#&gt; 4 AT2G22960.1 0.701       23 0.790       23 0.948       15 0.891 0.844</span>
<span class="co">#&gt; 5 AT2G19580.1 0.422       26 0.586       26 0.885       17 0.798 0.671</span>
<span class="co">#&gt; 6 AT2G19690.2 0.797       29 0.870       29 0.987       18 0.952 0.914</span>
<span class="co">#&gt;   is.sp Dmaxcut Networks.used is.signalp</span>
<span class="co">#&gt; 1     N   0.450  SignalP-noTM      FALSE</span>
<span class="co">#&gt; 2     Y   0.450  SignalP-noTM       TRUE</span>
<span class="co">#&gt; 3     Y   0.450  SignalP-noTM       TRUE</span>
<span class="co">#&gt; 4     Y   0.450  SignalP-noTM       TRUE</span>
<span class="co">#&gt; 5     Y   0.500    SignalP-TM       TRUE</span>
<span class="co">#&gt; 6     Y   0.450  SignalP-noTM       TRUE</span></code></pre></div>
<p>similarly <a href="http://www.cbs.dtu.dk/services/TargetP/">TargetP</a> and <a href="http://phobius.sbc.su.se/">Phobius</a> can be queried by:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">nsp_targetp &lt;-<span class="st"> </span><span class="kw"><a href="../../reference/get_targetp.html">get_targetp</a></span>(at_nsp,
                           sequence,
                           Transcript.id)

nsp_phobius &lt;-<span class="st"> </span><span class="kw"><a href="../../reference/get_phobius.html">get_phobius</a></span>(at_nsp,
                           sequence,
                           Transcript.id)</code></pre></div>
<p>The concordance of the predictions will be checked with Euler diagrams:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="dv">5</span>)
<span class="kw">bind_cols</span>(nsp_signalp, nsp_targetp, nsp_phobius) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">select</span>(is.phobius, is.targetp, is.signalp) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">euler</span>(<span class="dt">shape =</span> <span class="st">"ellipse"</span>, <span class="dt">input =</span> <span class="st">"disjoint"</span>) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span>(<span class="dt">quantities =</span> T)  </code></pre></div>
<center>
<div class="figure">
<img src="https://missuse.github.io/ragp/articles/eulerr.svg" width="240">
</div>
</center>
<p>The three servers do not agree completely. The disagreement is even higher when all Arabidopsis protein sequences are checked:</p>
<center>
<div class="figure">
<img src="https://missuse.github.io/ragp/articles/eulerr_all.svg" width="240">
</div>
</center>
<p>Therefore we recommend using the majority vote as the first filtering step:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">bind_cols</span>(nsp_signalp, nsp_targetp, nsp_phobius) <span class="op">%&gt;%</span><span class="st"> </span><span class="co">#bind the prediction data frames </span>
<span class="st">  </span><span class="kw">select</span>(is.phobius, is.targetp, is.signalp, id) <span class="op">%&gt;%</span><span class="st"> </span><span class="co">#select the logical columns from each, as well as the id</span>
<span class="st">  </span><span class="kw">mutate</span>(<span class="dt">vote =</span> <span class="kw"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span>(.[,<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>])) <span class="op">%&gt;%</span><span class="st"> </span><span class="co">#calculate the majority vote</span>
<span class="st">  </span><span class="kw"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span>(vote <span class="op">&gt;=</span><span class="st"> </span><span class="dv">2</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="co">#filter by majority vote</span>
<span class="st">  </span><span class="kw">pull</span>(id) -&gt;<span class="st"> </span>id_nsp <span class="co">#pull the filtered ids</span>

at_nsp_filtered &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span>(at_nsp,
                          Transcript.id <span class="op">%in%</span><span class="st"> </span>id_nsp) <span class="co">#filter at_nsp by pulled ids</span></code></pre></div>
<p>This will create <code>at_nsp_filtered</code> data frame with 2126 sequences.</p>
</div>
<div id="predicting-proline-hydroxylation" class="section level2">
<h2 class="hasAnchor">
<a href="#predicting-proline-hydroxylation" class="anchor"></a>Predicting proline hydroxylation</h2>
<p>The key feature of HRGPs is the presence of hydroxyprolines (Hyp, O) which represent glycosylation sites <span class="citation">(Showalter et al. 2010)</span>. While many HRGPs can be found based on biased amino acid composition, or the presence of certain amino acid motifs, there exists an abundance of chimeric proteins comprised from specific domains and HRGP motifs which are much harder to identify based on the mentioned features. In an attempt to identify these types of sequences <code>ragp</code> incorporates a model specifically built to predict the probability of proline hydroxylation in plant proteins. This model is incorporated in the <code><a href="../../reference/predict_hyp.html">predict_hyp()</a></code> function and can be utilized to filter potential HRGPs.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">at_hyp &lt;-<span class="st"> </span><span class="kw"><a href="../../reference/predict_hyp.html">predict_hyp</a></span>(<span class="dt">data =</span> at_nsp_filtered,
                      <span class="dt">sequence =</span> sequence,
                      <span class="dt">id =</span> Transcript.id)</code></pre></div>
<p>The object returned by <code>predict_hyp</code> is a list with two elements <code>prediction</code> and <code>sequence</code>. The <code>prediction</code> element is a data frame containing the probability of hydroxylation for each P in each sequence along with the amino acid span used for prediction. First few rows of this element:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(at_hyp<span class="op">$</span>prediction) </code></pre></div>
<pre><code>#&gt;            id                substr P_pos  prob HYP
#&gt; 1 AT2G43600.1 FSQNCMDTSCPGLKECCSRWG    31 0.015  No
#&gt; 2 AT2G43600.1 EYCGFFCFSGPCNIKGKSYGY    58 0.016  No
#&gt; 3 AT2G43600.1 YGYDYNVDAGPRGKIETVITS    76 0.018  No
#&gt; 4 AT2G43600.1 ERYCSKSKKYPCEPGKNYYGR   163 0.015  No
#&gt; 5 AT2G43600.1 CSKSKKYPCEPGKNYYGRGLL   166 0.014  No
#&gt; 6 AT2G43600.1 YYGAGKHLGLPLLKDPDLVSR   194 0.015  No</code></pre>
<p>The <code>sequence</code> (<code>at_hyp$sequence</code>) element is a data frame with the same sequences as provided in the function call in which prolines (P) are replaced by hydroxyprolines (O) at positions predicted to be hydroxylated. This element can be useful as input to <code><a href="../../reference/scan_ag.html">scan_ag()</a></code> as explained in <a href="https://missuse.github.io/ragp/articles/pkgdown/analyse">HRGP analysis</a> tutorial.</p>
<p>The predictions are based on a probability threshold of 0.224 which offers the best trade-off between specificity and sensitivity. To increase specificity at the cost of sensitivity the threshold can be increased using the <code>tprob</code> argument:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">at_hyp2 &lt;-<span class="st"> </span><span class="kw"><a href="../../reference/predict_hyp.html">predict_hyp</a></span>(<span class="dt">data =</span> at_nsp_filtered,
                       <span class="dt">sequence =</span> sequence,
                       <span class="dt">id =</span> Transcript.id,
                       <span class="dt">tprob =</span> <span class="fl">0.6</span>)</code></pre></div>
<p>The <code>at_hyp</code> object can be used to filter the sequences that contain more than a certain number of hydroxyprolines. The default threshold will be used. To filter sequences with three or more O:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">at_hyp<span class="op">$</span>prediction <span class="op">%&gt;%</span><span class="st"> </span><span class="co">#the prediction element</span>
<span class="st">  </span><span class="kw">group_by</span>(id) <span class="op">%&gt;%</span><span class="st"> </span><span class="co">#group by id</span>
<span class="st">  </span><span class="kw">summarise</span>(<span class="dt">n =</span> <span class="kw"><a href="https://rdrr.io/r/base/sum.html">sum</a></span>(HYP <span class="op">==</span><span class="st"> "Yes"</span>)) <span class="op">%&gt;%</span><span class="st"> </span><span class="co">#calculate the number of hydroxyprolines per sequence</span>
<span class="st">  </span><span class="kw"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span>(n <span class="op">&gt;=</span><span class="st"> </span><span class="dv">3</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="co">#filter sequences with three or more predicted hydroxyprolines</span>
<span class="st">  </span><span class="kw">pull</span>(id) -&gt;<span class="st"> </span>at_3hyp <span class="co">#pull the ids of the filtered sequences</span>

at_nsp_filtered <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span>(Transcript.id <span class="op">%in%</span><span class="st"> </span>at_3hyp) -&gt;<span class="st"> </span>at_nsp_3hyp <span class="co">#filter sequences from `at_nsp_filtered`</span></code></pre></div>
<p>There are 398 sequences that satisfy this condition.</p>
<p>After the filtering step the sequences can be analyzed by <code><a href="../../reference/scan_ag.html">scan_ag()</a></code> to identify arabinogalactan motifs, by <code><a href="../../reference/maab.html">maab()</a></code> to classify prototypical HRGPs, by <code><a href="../../reference/get_big_pi.html">get_big_pi()</a></code> and <code><a href="../../reference/get_pred_gpi.html">get_pred_gpi()</a></code> to predict GPI anchor sites in sequences and by <code><a href="../../reference/get_hmm.html">get_hmm()</a></code> and <code><a href="../../reference/get_espritz.html">get_espritz()</a></code> to identify domains and disordered regions. This is explained in <a href="https://missuse.github.io/ragp/articles/pkgdown/analyse">HRGP analysis</a> tutorial.</p>
</div>
<div id="references" class="section level2 unnumbered">
<h2 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h2>
<div id="refs" class="references">
<div id="ref-emanuelsson_predicting_2000">
<p>Emanuelsson, O., H. Nielsen, S. Brunak, and G. von Heijne. 2000. “Predicting Subcellular Localization of Proteins Based on Their N-Terminal Amino Acid Sequence.” <em>Journal of Molecular Biology</em> 300 (4): 1005–16. doi:<a href="https://doi.org/10.1006/jmbi.2000.3903">10.1006/jmbi.2000.3903</a>.</p>
</div>
<div id="ref-kall_advantages_2007">
<p>Käll, Lukas, Anders Krogh, and Erik L. L. Sonnhammer. 2007. “Advantages of Combined Transmembrane Topology and Signal Peptide Prediction–the Phobius Web Server.” <em>Nucleic Acids Research</em> 35 (Web Server issue): W429–432. doi:<a href="https://doi.org/10.1093/nar/gkm256">10.1093/nar/gkm256</a>.</p>
</div>
<div id="ref-showalter_bioinformatics_2010">
<p>Showalter, Allan M., Brian Keppler, Jens Lichtenberg, Dazhang Gu, and Lonnie R. Welch. 2010. “A Bioinformatics Approach to the Identification, Classification, and Analysis of Hydroxyproline-Rich Glycoproteins.” <em>Plant Physiology</em> 153 (2): 485–513. doi:<a href="https://doi.org/10.1104/pp.110.156554">10.1104/pp.110.156554</a>.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#predicting-n-terminal-secretory-signal-sequences-n-sp">Predicting N-terminal secretory signal sequences (N-sp)</a></li>
      <li><a href="#predicting-proline-hydroxylation">Predicting proline hydroxylation</a></li>
      <li><a href="#references">References</a></li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Milan Dragicevic, Danijela Paunovic.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.0.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
