<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Filtering hydroxyproline rich glycoproteins • ragp</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../../bootstrap-toc.css">
<script src="../../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../../pkgdown.css" rel="stylesheet">
<script src="../../pkgdown.js"></script><meta property="og:title" content="Filtering hydroxyproline rich glycoproteins">
<meta property="og:description" content="ragp">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-inverse navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../../index.html">ragp</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.3.5.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../../articles/pkgdown/start.html">Getting started</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../../articles/pkgdown/filter.html">Filtering hydroxyproline rich glycoproteins</a>
    </li>
    <li>
      <a href="../../articles/pkgdown/analyse.html">Analyse hydroxyproline rich glycoproteins</a>
    </li>
    <li>
      <a href="../../articles/pkgdown/PSV.html">Protein Sequence Visualization</a>
    </li>
  </ul>
</li>
<li>
  <a href="../../reference/index.html">Functions</a>
</li>
<li>
  <a href="../../news/index.html">News</a>
</li>
<li>
  <a href="https://ragp.shinyapps.io/Rapp/">ragp shiny</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/missuse/ragp/">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="filter_files/header-attrs-2.11/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Filtering hydroxyproline rich glycoproteins</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/missuse/ragp/blob/master/vignettes/pkgdown/filter.Rmd"><code>vignettes/pkgdown/filter.Rmd</code></a></small>
      <div class="hidden name"><code>filter.Rmd</code></div>

    </div>

    
    
<style type="text/css">

body{ /* Normal  */
      font-size: 15px;
  }
code.r{ /* Code block */
    font-size: 14px;
}
  
</style>
<p>The filtering layer in <code>ragp</code> consists of two sequential tasks:</p>
<ul>
<li>predict the presence of secretory signals (N-sp) and filter sequences containing them.<br>
</li>
<li>predict proline hydroxylation and filter sequences containing at least several potential hydroxyprolines.</li>
</ul>
<p>The following section explains how to perform these tasks on a set of 2700 Arabidopsis protein sequences included in <code>ragp</code> as `<code><a href="../../reference/at_nsp.html">at_nsp()</a></code> data frame.</p>
<p>Required packages:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/message.html">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/missuse/ragp">ragp</a></span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/message.html">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div id="predicting-n-terminal-secretory-signal-sequences-n-sp" class="section level2">
<h2 class="hasAnchor">
<a href="#predicting-n-terminal-secretory-signal-sequences-n-sp" class="anchor"></a>Predicting N-terminal secretory signal sequences (N-sp)</h2>
<p>Hydroxyproline rich glycoproteins (HRGPs) are secreted proteins, therefore they are expected to contain a secretory signal sequence on the N-terminus (N-sp). <code>ragp</code> incorporates N-sp prediction by querying <a href="https://services.healthtech.dtu.dk/service.php?SignalP-5.0">SignalP 5</a> <span class="citation">(Almagro Armenteros et al. 2019)</span>, <a href="https://services.healthtech.dtu.dk/service.php?SignalP-4.1">SignalP 4.1</a>, <a href="https://services.healthtech.dtu.dk/service.php?TargetP-1.1">TargetP 1.1</a> <span class="citation">(Emanuelsson et al. 2000)</span> and <a href="http://phobius.sbc.su.se/">Phobius</a> <span class="citation">(Käll, Krogh, and Sonnhammer 2007)</span> via the functions: <code><a href="../../reference/get_signalp5.html">get_signalp5()</a></code>, <code><a href="../../reference/get_signalp.html">get_signalp()</a></code>, <code><a href="../../reference/get_targetp.html">get_targetp()</a></code> and <code><a href="../../reference/get_phobius.html">get_phobius()</a></code>.</p>
<p>In the <code>ragp</code> versions prior <code>0.3.5</code> we advised a majority vote between <a href="https://services.healthtech.dtu.dk/service.php?SignalP-4.1">SignalP 4.1</a>, <a href="https://services.healthtech.dtu.dk/service.php?TargetP-1.1">TargetP 1.1</a> and <a href="http://phobius.sbc.su.se/">Phobius</a> should be used to determine which sequences should be flagged as secreted. Since the newest SignalP (version 5) utilizes a more sophisticated prediction method and a larger training set compared to the older methods we trust it provides a more accurate prediction of secreted protein sequences compared to the before mentioned majority vote.</p>
<p>To query <a href="https://services.healthtech.dtu.dk/service.php?SignalP-5.0">SignalP 5</a> predictions:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">nsp_signalp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/get_signalp5.html">get_signalp5</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">at_nsp</span>, <span class="co">#data frame name</span>
                            sequence <span class="op">=</span> <span class="va">sequence</span>, <span class="co">#column containing the sequences</span>
                            id <span class="op">=</span> <span class="va">Transcript.id</span><span class="op">)</span> <span class="co">#column containing protein identifiers</span></code></pre></div>
<p>The predictions for the 2700 sequences contained in <code>at_nsp</code> data frame should be available after several minutes. The returned object <code>nsp_signalp</code> is a data frame resembling the web servers output:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">nsp_signalp</span><span class="op">)</span>
<span class="co">#&gt;            id  Prediction SP.Sec.SPI  Other CS_pos   Pr cleave.site is.signalp sp.length</span>
<span class="co">#&gt; 1 ATCG00660.1       OTHER    0.00038 0.9996          NA                  FALSE        NA</span>
<span class="co">#&gt; 2 AT2G43600.1 SP(Sec/SPI)    0.99980 0.0002  22-23 0.96      VFS-QN       TRUE        22</span>
<span class="co">#&gt; 3 AT2G28410.1 SP(Sec/SPI)    0.99042 0.0096  22-23 0.89      ALA-QD       TRUE        22</span>
<span class="co">#&gt; 4 AT2G22960.1 SP(Sec/SPI)    0.99814 0.0019  22-23 0.94      AES-GS       TRUE        22</span>
<span class="co">#&gt; 5 AT2G19580.1       OTHER    0.26479 0.7352          NA                  FALSE        NA</span>
<span class="co">#&gt; 6 AT2G19690.2 SP(Sec/SPI)    0.98954 0.0105  28-29 0.90      ARS-EE       TRUE        28</span></code></pre></div>
<p>To filter the sequences predicted to contain an N-sp:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">nsp_signalp</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">is.signalp</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="co">#filter rows where is.signalp column is TRUE</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html">pull</a></span><span class="op">(</span><span class="va">id</span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">id_nsp</span> <span class="co">#pull the id column into a vector</span>

<span class="va">at_nsp_filtered</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">at_nsp</span>,
                          <span class="va">Transcript.id</span> <span class="op">%in%</span> <span class="va">id_nsp</span><span class="op">)</span> <span class="co">#filter at_nsp by pulled ids</span></code></pre></div>
<p>This will create <code>at_nsp_filtered</code> data frame with 1890 sequences.</p>
</div>
<div id="predicting-proline-hydroxylation" class="section level2">
<h2 class="hasAnchor">
<a href="#predicting-proline-hydroxylation" class="anchor"></a>Predicting proline hydroxylation</h2>
<p>The key feature of HRGPs is the presence of hydroxyprolines (Hyp, O) which represent glycosylation sites <span class="citation">(Showalter et al. 2010)</span>. While many HRGPs can be found based on biased amino acid composition, or the presence of certain amino acid motifs, there exists an abundance of chimeric proteins comprised from specific domains and HRGP motifs which are much harder to identify based on the mentioned features. In an attempt to identify these types of sequences <code>ragp</code> incorporates a model specifically built to predict the probability of proline hydroxylation in plant proteins. This model is incorporated in the <code><a href="../../reference/predict_hyp.html">predict_hyp()</a></code> function and can be utilized to filter potential HRGPs.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">at_hyp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/predict_hyp.html">predict_hyp</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">at_nsp_filtered</span>,
                      sequence <span class="op">=</span> <span class="va">sequence</span>,
                      id <span class="op">=</span> <span class="va">Transcript.id</span><span class="op">)</span></code></pre></div>
<p>The object returned by <code>predict_hyp</code> is a list with two elements <code>prediction</code> and <code>sequence</code>. The <code>prediction</code> element is a data frame containing the probability of hydroxylation for each P in each sequence along with the amino acid span used for prediction. First few rows of this element:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">at_hyp</span><span class="op">$</span><span class="va">prediction</span><span class="op">)</span> </code></pre></div>
<pre><code>#&gt;            id                substr P_pos  prob HYP
#&gt; 1 AT2G43600.1 FSQNCMDTSCPGLKECCSRWG    31 0.015  No
#&gt; 2 AT2G43600.1 EYCGFFCFSGPCNIKGKSYGY    58 0.016  No
#&gt; 3 AT2G43600.1 YGYDYNVDAGPRGKIETVITS    76 0.018  No
#&gt; 4 AT2G43600.1 ERYCSKSKKYPCEPGKNYYGR   163 0.015  No
#&gt; 5 AT2G43600.1 CSKSKKYPCEPGKNYYGRGLL   166 0.014  No
#&gt; 6 AT2G43600.1 YYGAGKHLGLPLLKDPDLVSR   194 0.015  No</code></pre>
<p>The <code>sequence</code> (<code>at_hyp$sequence</code>) element is a data frame with the same sequences as provided in the function call in which prolines (P) are replaced by hydroxyprolines (O) at positions predicted to be hydroxylated. This element can be useful as input to <code><a href="../../reference/scan_ag.html">scan_ag()</a></code> as explained in <a href="https://missuse.github.io/ragp/articles/pkgdown/analyse">HRGP analysis</a> tutorial.</p>
<p>The predictions are based on a probability threshold of 0.224 which offers the best trade-off between specificity and sensitivity. To increase specificity at the cost of sensitivity the threshold can be increased using the <code>tprob</code> argument:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">at_hyp2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/predict_hyp.html">predict_hyp</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">at_nsp_filtered</span>,
                       sequence <span class="op">=</span> <span class="va">sequence</span>,
                       id <span class="op">=</span> <span class="va">Transcript.id</span>,
                       tprob <span class="op">=</span> <span class="fl">0.6</span><span class="op">)</span></code></pre></div>
<p>The <code>at_hyp</code> object can be used to filter the sequences that contain more than a certain number of hydroxyprolines. The default threshold will be used. To filter sequences with three or more O:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">at_hyp</span><span class="op">$</span><span class="va">prediction</span> <span class="op">%&gt;%</span> <span class="co">#the prediction element</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">id</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="co">#group by id</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarise</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">HYP</span> <span class="op">==</span> <span class="st">"Yes"</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="co">#calculate the number of hydroxyprolines per sequence</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">n</span> <span class="op">&gt;=</span> <span class="fl">3</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="co">#filter sequences with three or more predicted hydroxyprolines</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html">pull</a></span><span class="op">(</span><span class="va">id</span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">at_3hyp</span> <span class="co">#pull the ids of the filtered sequences</span>

<span class="va">at_nsp_filtered</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">Transcript.id</span> <span class="op">%in%</span> <span class="va">at_3hyp</span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">at_nsp_3hyp</span> <span class="co">#filter sequences from `at_nsp_filtered`</span></code></pre></div>
<p>There are 365 sequences that satisfy this condition.</p>
<p>After the filtering step the sequences can be analyzed by <code><a href="../../reference/scan_ag.html">scan_ag()</a></code> to identify arabinogalactan motifs, by <code><a href="../../reference/maab.html">maab()</a></code> to classify prototypical HRGPs, by <code><a href="../../reference/get_big_pi.html">get_big_pi()</a></code>, <code><a href="../../reference/get_pred_gpi.html">get_pred_gpi()</a></code> or <code><a href="../../reference/get_netGPI.html">get_netGPI()</a></code> to predict GPI anchor sites in sequences and by <code><a href="../../reference/get_hmm.html">get_hmm()</a></code>, <code><a href="../../reference/get_cdd.html">get_cdd()</a></code> and <code><a href="../../reference/get_espritz.html">get_espritz()</a></code> to identify domains and disordered regions. This is explained in <a href="https://missuse.github.io/ragp/articles/pkgdown/analyse">HRGP analysis</a> tutorial.</p>
</div>
<div id="references" class="section level2 unnumbered">
<h2 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-armenteros_signalp5_2019" class="csl-entry">
Almagro Armenteros, José Juan, Konstantinos D. Tsirigos, Casper Kaae Sønderby, Thomas Nordahl Petersen, Ole Winther, Søren Brunak, Gunnar von Heijne, and Henrik Nielsen. 2019. <span>“SignalP 5.0 Improves Signal Peptide Predictions Using Deep Neural Networks.”</span> <em>Nature Biotechnology</em> 37: 420–23. <a href="https://doi.org/0.1038/s41587-019-0036-z">https://doi.org/0.1038/s41587-019-0036-z</a>.
</div>
<div id="ref-emanuelsson_predicting_2000" class="csl-entry">
Emanuelsson, O., H. Nielsen, S. Brunak, and G. von Heijne. 2000. <span>“Predicting Subcellular Localization of Proteins Based on Their <span>N</span>-Terminal Amino Acid Sequence.”</span> <em>Journal of Molecular Biology</em> 300 (4): 1005–16. <a href="https://doi.org/10.1006/jmbi.2000.3903">https://doi.org/10.1006/jmbi.2000.3903</a>.
</div>
<div id="ref-kall_advantages_2007" class="csl-entry">
Käll, Lukas, Anders Krogh, and Erik L. L. Sonnhammer. 2007. <span>“Advantages of Combined Transmembrane Topology and Signal Peptide Prediction–the <span>Phobius</span> Web Server.”</span> <em>Nucleic Acids Research</em> 35 (Web Server issue): W429–432. <a href="https://doi.org/10.1093/nar/gkm256">https://doi.org/10.1093/nar/gkm256</a>.
</div>
<div id="ref-showalter_bioinformatics_2010" class="csl-entry">
Showalter, Allan M., Brian Keppler, Jens Lichtenberg, Dazhang Gu, and Lonnie R. Welch. 2010. <span>“A <span>Bioinformatics</span> <span>Approach</span> to the <span>Identification</span>, <span>Classification</span>, and <span>Analysis</span> of <span>Hydroxyproline</span>-<span>Rich</span> <span>Glycoproteins</span>.”</span> <em>Plant Physiology</em> 153 (2): 485–513. <a href="https://doi.org/10.1104/pp.110.156554">https://doi.org/10.1104/pp.110.156554</a>.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Milan Dragicevic, Danijela Paunovic.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
