<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Protein Sequence Visualization • ragp</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../../bootstrap-toc.css">
<script src="../../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../../pkgdown.css" rel="stylesheet">
<script src="../../pkgdown.js"></script><meta property="og:title" content="Protein Sequence Visualization">
<meta property="og:description" content="ragp">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-inverse navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../../index.html">ragp</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.3.5.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../../articles/pkgdown/start.html">Getting started</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../../articles/pkgdown/filter.html">Filtering hydroxyproline rich glycoproteins</a>
    </li>
    <li>
      <a href="../../articles/pkgdown/analyse.html">Analyse hydroxyproline rich glycoproteins</a>
    </li>
    <li>
      <a href="../../articles/pkgdown/PSV.html">Protein Sequence Visualization</a>
    </li>
  </ul>
</li>
<li>
  <a href="../../reference/index.html">Functions</a>
</li>
<li>
  <a href="../../news/index.html">News</a>
</li>
<li>
  <a href="https://ragp.shinyapps.io/Rapp/">ragp shiny</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/missuse/ragp/">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="PSV_files/header-attrs-2.11/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Protein Sequence Visualization</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/missuse/ragp/blob/master/vignettes/pkgdown/PSV.Rmd"><code>vignettes/pkgdown/PSV.Rmd</code></a></small>
      <div class="hidden name"><code>PSV.Rmd</code></div>

    </div>

    
    
<style type="text/css">

body{ /* Normal  */
      font-size: 15px;
  }
code.r{ /* Code block */
    font-size: 14px;
}
  
</style>
<p>All of the protein sequence attributes which can be acquired using <code>ragp</code> functions can be visualized using the function <code><a href="../../reference/plot_prot.html">plot_prot()</a></code> which accepts protein sequences in the form of strings, calls the appropriate <code>ragp</code> functions to obtain sequence features and returns a <code>ggplot2</code> object.</p>
<p>The following section explains how to use <code><a href="../../reference/plot_prot.html">plot_prot()</a></code> and manipulate the output visual style.</p>
<p>Required packages:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://nanx.me/protr/">protr</a></span><span class="op">)</span> 
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/missuse/ragp">ragp</a></span><span class="op">)</span></code></pre></div>
<div id="basic-usage" class="section level2">
<h2 class="hasAnchor">
<a href="#basic-usage" class="anchor"></a>Basic usage</h2>
<p>In the next example the function <code><a href="https://nanx.me/protr/reference/getUniProt.html">protr::getUniProt()</a></code> will be used to download several protein sequences from <a href="https://www.uniprot.org/">UniProt</a> and then <code><a href="../../reference/plot_prot.html">plot_prot()</a></code> will be used to visualize the sequence attributes. Do note that downloading the sequences from UniProt is not a requirement for <code><a href="../../reference/plot_prot.html">plot_prot()</a></code>. No information apart the actual protein sequences and their identifiers (accessions) is required for <code><a href="../../reference/plot_prot.html">plot_prot()</a></code>:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Q9C7T7"</span>,
         <span class="st">"Q9ZPR7"</span>,
         <span class="st">"Q9ZQ23"</span><span class="op">)</span></code></pre></div>
<p>Download the corresponding sequences using <code><a href="https://nanx.me/protr/reference/getUniProt.html">protr::getUniProt()</a></code> and unlist them so they are in an appropriate format for <code><a href="../../reference/plot_prot.html">plot_prot()</a></code>:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">seqs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span><span class="op">(</span><span class="fu">protr</span><span class="fu">::</span><span class="fu"><a href="https://nanx.me/protr/reference/getUniProt.html">getUniProt</a></span><span class="op">(</span><span class="va">ids</span><span class="op">)</span><span class="op">)</span>
<span class="va">seqs</span>
<span class="co">#&gt; [1] "MSRRPDLLRGSVVATVAATFLLFIFPPNVESTVEKQALFRFKNRLDDSHNILQSWKPSDSPCVFRGITCDPLSGEVIGISLGNVNLSGTISPSISALTKLSTLSLPSNFISGRIPPEIVNCKNLKVLNLTSNRLSGTIPNLSPLKSLEILDISGNFLNGEFQSWIGNMNQLVSLGLGNNHYEEGIIPESIGGLKKLTWLFLARSNLTGKIPNSIFDLNALDTFDIANNAISDDFPILISRLVNLTKIELFNNSLTGKIPPEIKNLTRLREFDISSNQLSGVLPEELGVLKELRVFHCHENNFTGEFPSGFGDLSHLTSLSIYRNNFSGEFPVNIGRFSPLDTVDISENEFTGPFPRFLCQNKKLQFLLALQNEFSGEIPRSYGECKSLLRLRINNNRLSGQVVEGFWSLPLAKMIDLSDNELTGEVSPQIGLSTELSQLILQNNRFSGKIPRELGRLTNIERIYLSNNNLSGEIPMEVGDLKELSSLHLENNSLTGFIPKELKNCVKLVDLNLAKNFLTGEIPNSLSQIASLNSLDFSGNRLTGEIPASLVKLKLSFIDLSGNQLSGRIPPDLLAVGGSTAFSRNEKLCVDKENAKTNQNLGLSICSGYQNVKRNSSLDGTLLFLALAIVVVVLVSGLFALRYRVVKIRELDSENRDINKADAKWKIASFHQMELDVDEICRLDEDHVIGSGSAGKVYRVDLKKGGGTVAVKWLKRGGGEEGDGTEVSVAEMEILGKIRHRNVLKLYACLVGRGSRYLVFEFMENGNLYQALGNNIKGGLPELDWLKRYKIAVGAAKGIAYLHHDCCPPIIHRDIKSSNILLDGDYESKIADFGVAKVADKGYEWSCVAGTHGYMAPELAYSFKATEKSDVYSFGVVLLELVTGLRPMEDEFGEGKDIVDYVYSQIQQDPRNLQNVLDKQVLSTYIEESMIRVLKMGLLCTTKLPNLRPSMREVVRKLDDADPCVSNSQDTTGKITV"</span>
<span class="co">#&gt; [2] "MYMIESKGGAIACMLLALLFLGTWPAIMTLTERRGRLPQHTYLDYTLTNLLAAVIIALTLGEIGPSRPNFFTQLSQDNWQSVMFAMAGGIVLSLGNLATQYAWAYVGLSVTEVITASITVVIGTTLNYFLDDRINRAEVLFPGVACFLIAVCFGSAVHKSNAADNKTKLQNFKSLETTSSFEMETISASNGLTKGKAKEGTAAFLIELEKQRAIKVFGKSTIIGLVITFFAGICFSLFSPAFNLATNDQWHTLKHGVPKLNVYTAFFYFSISAFVVALILNIRFLYWPILGLPRSSFKAYLNDWNGRGWSFLAGFLCGFGNGLQFMGGQAAGYAAADAVQALPLVSTFWGILLFGEYRRSSRKTYTLLISMLLMFIVAVAVLMASSGHRK"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           </span>
<span class="co">#&gt; [3] "MGLKVSSSLLCLTILLAVSSIVSAVNITRVLEKYPEFSTMTELLAKTELTPIINKRQTITVLALNNDAIGSISGRPEEEVKNILMNHVVLDYFDELKLKALKEKSTLLTTLYQSTGLGQQQNGFLNCTKSNGKIYFGSGVKGAPQTAEYITTVFRNPYNLSVVQISMPIVAPGLGSPVKVPPPPPMSSPPAPSPKKGAATPAPAPADEGDYADAPPGLAPETAPASAPSESDSPAPAPDKSGKKKMAAADEAEPPSSASNTGLSFGAVLVLGFVASFVGF"</span></code></pre></div>
<p>Provide the sequences and protein identifiers to <code><a href="../../reference/plot_prot.html">plot_prot()</a></code>:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/plot_prot.html">plot_prot</a></span><span class="op">(</span>sequence <span class="op">=</span> <span class="va">seqs</span>,
                id <span class="op">=</span> <span class="va">ids</span><span class="op">)</span>
<span class="va">p1</span></code></pre></div>
<p><img src="https://missuse.github.io/ragp/articles/PSV_p1.svg"></p>
<p>By default the plot contains the following information:</p>
<ul>
<li><p>Transmembrane regions (TM) are shown in yellow. Extracellular regions are indicated by the dashed line above the sequences while intracellular regions are indicated by the dashed line bellow the sequence. For TM prediction <code><a href="../../reference/plot_prot.html">plot_prot()</a></code> can query <a href="http://phobius.sbc.su.se/">Phobius</a> or <a href="https://services.healthtech.dtu.dk/service.php?TMHMM-2.0">TMHMM</a> via the functions <code><a href="../../reference/get_phobius.html">get_phobius()</a></code> and <code><a href="../../reference/get_tmhmm.html">get_tmhmm()</a></code>. The <code><a href="../../reference/plot_prot.html">plot_prot()</a></code> argument <code>tm</code> (can be: “phobius”, “tmhmm” or “none”) instructs which method should be used. Default is “phobius”.</p></li>
<li><p>Signal peptides are indicated by the thick red line on the N-terminal side. For N-sp prediction <code><a href="../../reference/plot_prot.html">plot_prot()</a></code> can query <a href="https://services.healthtech.dtu.dk/service.php?SignalP-4.1">SignalP 4.1</a> or <a href="https://services.healthtech.dtu.dk/service.php?SignalP-5.0">SignalP-5.0</a>. The <code><a href="../../reference/plot_prot.html">plot_prot()</a></code> argument <code>nsp</code> (can be: “signalp”, “signalp5” or “none”) instructs which method should be used. Default is “signalp5”.</p></li>
<li><p>GPI attachment sites (as predicted by <code><a href="../../reference/get_big_pi.html">get_big_pi()</a></code>, <code><a href="../../reference/get_pred_gpi.html">get_pred_gpi()</a></code> or <code><a href="../../reference/get_netGPI.html">get_netGPI()</a></code> depending on the <code><a href="../../reference/plot_prot.html">plot_prot()</a></code> argument <code>gpi</code>) are indicated by blue rhomboids. Default is to query <a href="https://services.healthtech.dtu.dk/service.php?NetGPI-1.1">NetGPI</a></p></li>
<li><p>Hydroxyprolines (as predicted by <code><a href="../../reference/predict_hyp.html">predict_hyp()</a></code>) are indicated by vertical dark gray lines.</p></li>
<li><p>Arabinogalactan glycomodule spans are indicated by the light grey background (motifs are found using <code><a href="../../reference/scan_ag.html">scan_ag()</a></code>).</p></li>
<li><p>Domains are indicated by rectangles with an appropriate fill as indicated in the legend. Domain annotation for the plot can be obtained using <code><a href="../../reference/get_hmm.html">get_hmm()</a></code> or <code><a href="../../reference/get_cdd.html">get_cdd()</a></code> depending on the <code>domain</code> argument (can be: “cdd”, “hmm” or “none”). Default is to use <code><a href="../../reference/get_cdd.html">get_cdd()</a></code>.</p></li>
</ul>
<p>Plotting can take a while since <code><a href="../../reference/plot_prot.html">plot_prot()</a></code> first obtains the appropriate predictions using the above mentioned function and then generates the plot. To check what actions the function is performing set the argument <code>progress</code> to <code>TRUE</code>.</p>
<p>Alternatively the outputs of <code><a href="../../reference/get_hmm.html">get_hmm()</a></code>, <code><a href="../../reference/get_cdd.html">get_cdd()</a></code>, <code><a href="../../reference/get_signalp.html">get_signalp()</a></code>, <code><a href="../../reference/get_signalp5.html">get_signalp5()</a></code>, <code><a href="../../reference/get_phobius.html">get_phobius()</a></code>, <code><a href="../../reference/get_tmhmm.html">get_tmhmm()</a></code>, <code><a href="../../reference/get_big_pi.html">get_big_pi()</a></code>, <code><a href="../../reference/get_pred_gpi.html">get_pred_gpi()</a></code>, <code><a href="../../reference/get_netGPI.html">get_netGPI()</a></code> and <code><a href="../../reference/get_espritz.html">get_espritz()</a></code> can be supplied as <code><a href="../../reference/plot_prot.html">plot_prot()</a></code> arguments <code>domain</code>, <code>nsp</code>, <code>tm</code>, <code>gpi</code> and <code>disorder</code>:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">domains</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/get_cdd.html">get_cdd</a></span><span class="op">(</span>sequence <span class="op">=</span> <span class="va">seqs</span>,
                   id <span class="op">=</span> <span class="va">ids</span><span class="op">)</span>

<span class="va">tm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/get_phobius.html">get_phobius</a></span><span class="op">(</span>sequence <span class="op">=</span> <span class="va">seqs</span>,
                  id <span class="op">=</span> <span class="va">ids</span><span class="op">)</span>

<span class="va">nsp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/get_signalp5.html">get_signalp5</a></span><span class="op">(</span>sequence <span class="op">=</span> <span class="va">seqs</span>,
                   id <span class="op">=</span> <span class="va">ids</span><span class="op">)</span>

<span class="va">gpi</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/get_netGPI.html">get_netGPI</a></span><span class="op">(</span>sequence <span class="op">=</span> <span class="va">seqs</span>,
                  id <span class="op">=</span> <span class="va">ids</span><span class="op">)</span>

<span class="va">p1_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/plot_prot.html">plot_prot</a></span><span class="op">(</span>sequence <span class="op">=</span> <span class="va">seqs</span>,
                   id <span class="op">=</span> <span class="va">ids</span>,
                   domain <span class="op">=</span> <span class="va">domains</span>,
                   tm <span class="op">=</span> <span class="va">tm</span>,
                   nsp <span class="op">=</span> <span class="va">nsp</span>,
                   gpi <span class="op">=</span> <span class="va">gpi</span><span class="op">)</span>

<span class="va">p1_df</span></code></pre></div>
<p><img src="https://missuse.github.io/ragp/articles/PSV_p1.svg"> Mixing the two types of input is also a possibility. In the below call the data frames <code>nsp</code> and <code>gpi</code> are used to tell <code>prot_plot()</code> the length of the N-sp and the positions of omega sites while <code>domain = "hmm"</code> and <code>tm = "tmhmm"</code> instruct it to use <code><a href="../../reference/get_hmm.html">get_hmm()</a></code> to obtain domain predictions and <code><a href="../../reference/get_tmhmm.html">get_tmhmm()</a></code> to obtain TM predictions:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/plot_prot.html">plot_prot</a></span><span class="op">(</span>sequence <span class="op">=</span> <span class="va">seqs</span>,
                id <span class="op">=</span> <span class="va">ids</span>,
                domain <span class="op">=</span> <span class="st">"hmm"</span>,
                tm <span class="op">=</span> <span class="st">"tmhmm"</span>,
                nsp <span class="op">=</span> <span class="va">nsp</span>,
                gpi <span class="op">=</span> <span class="va">gpi</span><span class="op">)</span>
<span class="va">p2</span></code></pre></div>
<p><img src="https://missuse.github.io/ragp/articles/PSV_p2.svg"></p>
<p>From the above plot it can be observed TMHMM tends to predict TM regions where there are N-sp’s and that <code><a href="../../reference/get_cdd.html">get_cdd()</a></code> and <code><a href="../../reference/get_hmm.html">get_hmm()</a></code> differ in the type of detected domains. For instance the <a href="https://www.uniprot.org/uniprot/Q9C7T7">Q9C7T7</a> is a leucine-rich repeat (LRR) receptor-like protein kinase - the Conserved domain database has an entry for exactly the mentioned class <a href="https://www.ncbi.nlm.nih.gov/Structure/cdd/PLN00113">PLN00113</a> while the <a href="http://pfam.xfam.org/">Pfam database</a> has more general domain models so it predicts a protein kinase domain and several LRR domains.</p>
<p>To plot disordered regions as predicted by <a href="http://old.protein.bio.unipd.it/espritz/">Espritz</a> vie the function <code><a href="../../reference/get_espritz.html">get_espritz()</a></code>:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/plot_prot.html">plot_prot</a></span><span class="op">(</span>sequence <span class="op">=</span> <span class="va">seqs</span>,
                id <span class="op">=</span> <span class="va">ids</span>,
                domain <span class="op">=</span> <span class="va">domain</span>,
                tm <span class="op">=</span> <span class="va">tm</span>,
                nsp <span class="op">=</span> <span class="va">nsp</span>,
                gpi <span class="op">=</span> <span class="va">gpi</span>,
                disorder <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
                
<span class="va">p2</span></code></pre></div>
<p><img src="https://missuse.github.io/ragp/articles/PSV_p13.svg"></p>
<p>The longest predicted disordered region overlaps with the AG motif span in <a href="https://www.uniprot.org/uniprot/Q9ZQ23">Q9ZQ23</a></p>
</div>
<div id="plot_prot-options" class="section level2">
<h2 class="hasAnchor">
<a href="#plot_prot-options" class="anchor"></a><code>plot_prot()</code> options</h2>
<ul>
<li>
<strong>determining what features to plot</strong>: to turn off some elements from being plotted use the appropriate argument, for instance <code>hyp = FALSE</code> will result in hydroxyproline predictions to not be plotted (check full list of options: <code><a href="../../reference/plot_prot.html">?plot_prot</a></code>):</li>
</ul>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/plot_prot.html">plot_prot</a></span><span class="op">(</span>sequence <span class="op">=</span> <span class="va">seqs</span>,
                id <span class="op">=</span> <span class="va">ids</span>,
                hyp <span class="op">=</span> <span class="cn">FALSE</span>, <span class="co">#do not plot hydroxyprolines</span>
                ag <span class="op">=</span> <span class="cn">FALSE</span>, <span class="co">#do not plot arabinogalactan motifs</span>
                gpi <span class="op">=</span> <span class="st">"predgpi"</span>, <span class="co">#use `get_pred_gpi()` to obtain GPI predictions</span>
                domain <span class="op">=</span> <span class="st">"hmm"</span><span class="op">)</span>  <span class="co">#use `get_hmm()` to obtain domain predictions</span>
<span class="va">p1</span></code></pre></div>
<p><img src="https://missuse.github.io/ragp/articles/PSV_p3.svg"></p>
<ul>
<li>
<strong>feature specific options</strong>: passing appropriate function arguments to functions called by <code><a href="../../reference/plot_prot.html">plot_prot()</a></code> such as <code>dim</code> and <code>div</code> for <code><a href="../../reference/scan_ag.html">scan_ag()</a></code> or <code>tprob</code> for <code><a href="../../reference/predict_hyp.html">predict_hyp()</a></code> to change the rules of arabinogalactan motif scan and the threshold for decision if the proline is hydroxylated is supported:</li>
</ul>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/plot_prot.html">plot_prot</a></span><span class="op">(</span>sequence <span class="op">=</span> <span class="va">seqs</span>,
                id <span class="op">=</span> <span class="va">ids</span>,
                dim <span class="op">=</span> <span class="fl">5</span>, <span class="co">#`scan_ag()` argument, minimum number of dipeptides for a motif</span>
                div <span class="op">=</span> <span class="fl">5</span>, <span class="co">#`scan_ag()` argument, maximum number of amino acids between dipeptides </span>
                tprob <span class="op">=</span> <span class="fl">0.9</span>,
                domain <span class="op">=</span> <span class="st">"hmm"</span>,
                ievalue <span class="op">=</span> <span class="fl">1e-10</span><span class="op">)</span> <span class="co">#`get_hmm()` argument, remove domains with independent E-value higher than this value</span>
                
<span class="va">p2</span></code></pre></div>
<p><img src="https://missuse.github.io/ragp/articles/PSV_p4.svg"></p>
<ul>
<li>
<strong>change feature colors</strong>:</li>
</ul>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/plot_prot.html">plot_prot</a></span><span class="op">(</span>sequence <span class="op">=</span> <span class="va">seqs</span>,
                id <span class="op">=</span> <span class="va">ids</span>,
                hyp_col <span class="op">=</span> <span class="st">"chocolate2"</span>,
                gpi_col <span class="op">=</span> <span class="st">"lightsalmon4"</span>,
                nsp_col <span class="op">=</span> <span class="st">"sienna1"</span>,
                ag_col <span class="op">=</span> <span class="st">"bisque2"</span>,
                tm_col <span class="op">=</span> <span class="st">"aquamarine"</span><span class="op">)</span> 
<span class="va">p3</span></code></pre></div>
<p><img src="https://missuse.github.io/ragp/articles/PSV_p5.svg"></p>
<ul>
<li>
<strong>change the order of domain plotting</strong>: in the above figures where <code><a href="../../reference/plot_prot.html">plot_prot()</a></code> was called with <code>domain = "hmm"</code> it can be observed there are more entries in the legend then can be seen in the plot. Some domains cover other domains. By default the domains with the lowest I-evalue (independent e-value) are plotted on top. To change this the argument <code>dom_sort</code> can be used. This argument can take the following values: <code>"ievalue"</code> (default), <code>"abc"</code> (alphabetically sorted by domain names), <code>"cba"</code> (reverse of <code>"abc"</code>).</li>
</ul>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p4</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/plot_prot.html">plot_prot</a></span><span class="op">(</span>sequence <span class="op">=</span> <span class="va">seqs</span>,
                id <span class="op">=</span> <span class="va">ids</span>,
                dom_sort <span class="op">=</span> <span class="st">"cba"</span>,
                domain <span class="op">=</span> <span class="st">"hmm"</span><span class="op">)</span>
<span class="va">p4</span></code></pre></div>
<p><img src="https://missuse.github.io/ragp/articles/PSV_p6.svg"></p>
<p>Ureide permease domain overlaps with TMEM144 transmembrane transporter domain and also protein kinase and protein tyrosine kinase domains overlap.</p>
</div>
<div id="customizing-the-plot-using-ggplot-layers" class="section level2">
<h2 class="hasAnchor">
<a href="#customizing-the-plot-using-ggplot-layers" class="anchor"></a>Customizing the plot using ggplot layers</h2>
<p><code><a href="../../reference/plot_prot.html">plot_prot()</a></code> returns a <code>ggplot</code> object which can be further customized:</p>
<ul>
<li>
<strong>change theme elements</strong>:</li>
</ul>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p1</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span>,
        legend.direction <span class="op">=</span> <span class="st">"vertical"</span><span class="op">)</span></code></pre></div>
<p><img src="https://missuse.github.io/ragp/articles/PSV_p7.svg"></p>
<ul>
<li>
<strong>add annotations</strong>:</li>
</ul>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p1</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span>,
        legend.direction <span class="op">=</span> <span class="st">"vertical"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/annotate.html">annotate</a></span><span class="op">(</span>geom <span class="op">=</span> <span class="st">"text"</span>,
           y <span class="op">=</span> <span class="fl">3</span>,
           x <span class="op">=</span> <span class="fl">90</span>,
           label <span class="op">=</span> <span class="st">"FAS domain"</span><span class="op">)</span></code></pre></div>
<p><img src="https://missuse.github.io/ragp/articles/PSV_p8.svg"></p>
<p>The <code>y</code> axis is continuous - <code>1</code> on the plot corresponds to Q9C7T7, <code>2</code> corresponds to Q9ZPR7 and 3 to Q9ZQ23. The <code>x</code> axis represents amino acids in the sequence. This enables adding annotations anywhere on the plot.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p1</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span>,
        legend.direction <span class="op">=</span> <span class="st">"vertical"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/annotate.html">annotate</a></span><span class="op">(</span>geom <span class="op">=</span> <span class="st">"text"</span>,
           y <span class="op">=</span> <span class="fl">1.6</span>,
           x <span class="op">=</span> <span class="fl">530</span>,
           label <span class="op">=</span> <span class="st">"annotation 1"</span><span class="op">)</span></code></pre></div>
<p><img src="https://missuse.github.io/ragp/articles/PSV_p9.svg"></p>
<ul>
<li>
<strong>change plot ratio</strong>: the ratio of the axes is determined automatically by <code><a href="../../reference/plot_prot.html">plot_prot()</a></code> and scales with the longest sequence plotted. To change this use the <code>ratio</code> argument to <code><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html">coord_equal()</a></code>:</li>
</ul>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p1</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span>,
        legend.direction <span class="op">=</span> <span class="st">"vertical"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html">coord_equal</a></span><span class="op">(</span>ratio <span class="op">=</span> <span class="fl">80</span>,
              expand <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
<p><img src="https://missuse.github.io/ragp/articles/PSV_p10.svg"></p>
<ul>
<li>
<strong>change domain colors</strong>: use another fill palette:</li>
</ul>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p1</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span>,
        legend.direction <span class="op">=</span> <span class="st">"vertical"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html">scale_fill_brewer</a></span><span class="op">(</span>type <span class="op">=</span> <span class="st">"div"</span><span class="op">)</span></code></pre></div>
<p><img src="https://missuse.github.io/ragp/articles/PSV_p11.svg"></p>
<p>or create a manual palette:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p1</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span>,
        legend.direction <span class="op">=</span> <span class="st">"vertical"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_fill_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"orange"</span>,
                               <span class="st">"red"</span>,
                               <span class="st">"yellow"</span>,
                               <span class="st">"dodgerblue"</span>,
                               <span class="st">"green"</span>,
                               <span class="st">"blue"</span>,
                               <span class="st">"purple"</span>,
                               <span class="st">"grey30"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="https://missuse.github.io/ragp/articles/PSV_p12.svg"></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Milan Dragicevic, Danijela Paunovic.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
