<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Protein Sequence Visualization â€¢ ragp</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../../pkgdown.css" rel="stylesheet">
<script src="../../pkgdown.js"></script><meta property="og:title" content="Protein Sequence Visualization">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-inverse navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../../index.html">ragp</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.3.0.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../../articles/pkgdown/start.html">Getting started</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../../articles/pkgdown/filter.html">Filtering hydroxyproline rich glycoproteins</a>
    </li>
    <li>
      <a href="../../articles/pkgdown/analyse.html">Analyse hydroxyproline rich glycoproteins</a>
    </li>
    <li>
      <a href="../../articles/pkgdown/PSV.html">Protein Sequence Visualization</a>
    </li>
  </ul>
</li>
<li>
  <a href="../../reference/index.html">Functions</a>
</li>
<li>
  <a href="../../news/index.html">News</a>
</li>
<li>
  <a href="https://ragp.shinyapps.io/Rapp/">ragp shiny</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/missuse/ragp/">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Protein Sequence Visualization</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/missuse/ragp/blob/master/vignettes/pkgdown/PSV.Rmd"><code>vignettes/pkgdown/PSV.Rmd</code></a></small>
      <div class="hidden name"><code>PSV.Rmd</code></div>

    </div>

    
    
<style type="text/css">

body{ /* Normal  */
      font-size: 15px;
  }
code.r{ /* Code block */
    font-size: 14px;
}
  
</style>
<p>All of the protein sequence attributes which can be acquired using <code>ragp</code> functions can be visualized using the function <code><a href="../../reference/plot_prot.html">plot_prot()</a></code> which accepts protein sequences in the form of strings, calls the appropriate <code>ragp</code> functions to obtain sequence features and returns a <code>ggplot2</code> object.</p>
<p>The following section explains how to use <code><a href="../../reference/plot_prot.html">plot_prot()</a></code> and manipulate the output visual style.</p>
<p>Required packages:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(ggplot2)
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(protr) 
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(ragp)</code></pre></div>
<div id="basic-usage" class="section level2">
<h2 class="hasAnchor">
<a href="#basic-usage" class="anchor"></a>Basic usage</h2>
<p>In the next example the function <code><a href="https://rdrr.io/pkg/protr/man/getUniProt.html">protr::getUniProt()</a></code> will be used to download several protein sequences from <a href="https://www.uniprot.org/">UniProt</a> and then <code><a href="../../reference/plot_prot.html">plot_prot()</a></code> will be used to visualize the sequence attributes. Do note that downloading the sequences from UniProt is not a requirement for <code><a href="../../reference/plot_prot.html">plot_prot()</a></code>. No information apart the actual protein sequences and their identifiers (accessions) is required for <code><a href="../../reference/plot_prot.html">plot_prot()</a></code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ids &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"Q9C7T7"</span>,
         <span class="st">"Q9ZPR7"</span>,
         <span class="st">"Q9ZQ23"</span>)</code></pre></div>
<p>Download the corresponding sequences using <code><a href="https://rdrr.io/pkg/protr/man/getUniProt.html">protr::getUniProt()</a></code> and unlist them so they are in an appropriate format for <code><a href="../../reference/plot_prot.html">plot_prot()</a></code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">seqs &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span>(protr<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/protr/man/getUniProt.html">getUniProt</a></span>(ids))
seqs
<span class="co">#&gt; [1] "MSRRPDLLRGSVVATVAATFLLFIFPPNVESTVEKQALFRFKNRLDDSHNILQSWKPSDSPCVFRGITCDPLSGEVIGISLGNVNLSGTISPSISALTKLSTLSLPSNFISGRIPPEIVNCKNLKVLNLTSNRLSGTIPNLSPLKSLEILDISGNFLNGEFQSWIGNMNQLVSLGLGNNHYEEGIIPESIGGLKKLTWLFLARSNLTGKIPNSIFDLNALDTFDIANNAISDDFPILISRLVNLTKIELFNNSLTGKIPPEIKNLTRLREFDISSNQLSGVLPEELGVLKELRVFHCHENNFTGEFPSGFGDLSHLTSLSIYRNNFSGEFPVNIGRFSPLDTVDISENEFTGPFPRFLCQNKKLQFLLALQNEFSGEIPRSYGECKSLLRLRINNNRLSGQVVEGFWSLPLAKMIDLSDNELTGEVSPQIGLSTELSQLILQNNRFSGKIPRELGRLTNIERIYLSNNNLSGEIPMEVGDLKELSSLHLENNSLTGFIPKELKNCVKLVDLNLAKNFLTGEIPNSLSQIASLNSLDFSGNRLTGEIPASLVKLKLSFIDLSGNQLSGRIPPDLLAVGGSTAFSRNEKLCVDKENAKTNQNLGLSICSGYQNVKRNSSLDGTLLFLALAIVVVVLVSGLFALRYRVVKIRELDSENRDINKADAKWKIASFHQMELDVDEICRLDEDHVIGSGSAGKVYRVDLKKGGGTVAVKWLKRGGGEEGDGTEVSVAEMEILGKIRHRNVLKLYACLVGRGSRYLVFEFMENGNLYQALGNNIKGGLPELDWLKRYKIAVGAAKGIAYLHHDCCPPIIHRDIKSSNILLDGDYESKIADFGVAKVADKGYEWSCVAGTHGYMAPELAYSFKATEKSDVYSFGVVLLELVTGLRPMEDEFGEGKDIVDYVYSQIQQDPRNLQNVLDKQVLSTYIEESMIRVLKMGLLCTTKLPNLRPSMREVVRKLDDADPCVSNSQDTTGKITV"</span>
<span class="co">#&gt; [2] "MYMIESKGGAIACMLLALLFLGTWPAIMTLTERRGRLPQHTYLDYTLTNLLAAVIIALTLGEIGPSRPNFFTQLSQDNWQSVMFAMAGGIVLSLGNLATQYAWAYVGLSVTEVITASITVVIGTTLNYFLDDRINRAEVLFPGVACFLIAVCFGSAVHKSNAADNKTKLQNFKSLETTSSFEMETISASNGLTKGKAKEGTAAFLIELEKQRAIKVFGKSTIIGLVITFFAGICFSLFSPAFNLATNDQWHTLKHGVPKLNVYTAFFYFSISAFVVALILNIRFLYWPILGLPRSSFKAYLNDWNGRGWSFLAGFLCGFGNGLQFMGGQAAGYAAADAVQALPLVSTFWGILLFGEYRRSSRKTYTLLISMLLMFIVAVAVLMASSGHRK"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           </span>
<span class="co">#&gt; [3] "MGLKVSSSLLCLTILLAVSSIVSAVNITRVLEKYPEFSTMTELLAKTELTPIINKRQTITVLALNNDAIGSISGRPEEEVKNILMNHVVLDYFDELKLKALKEKSTLLTTLYQSTGLGQQQNGFLNCTKSNGKIYFGSGVKGAPQTAEYITTVFRNPYNLSVVQISMPIVAPGLGSPVKVPPPPPMSSPPAPSPKKGAATPAPAPADEGDYADAPPGLAPETAPASAPSESDSPAPAPDKSGKKKMAAADEAEPPSSASNTGLSFGAVLVLGFVASFVGF"</span></code></pre></div>
<p>Provide the sequences and protein identifiers to <code><a href="../../reference/plot_prot.html">plot_prot()</a></code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">p1 &lt;-<span class="st"> </span><span class="kw"><a href="../../reference/plot_prot.html">plot_prot</a></span>(<span class="dt">sequence =</span> seqs,
                <span class="dt">id =</span> ids)
p1</code></pre></div>
<div class="figure">
<img src="https://missuse.github.io/ragp/articles/PSV_p1.svg">
</div>
<p>By default the plot contains the following information:</p>
<ul>
<li>Transmembrane regions (TM) are shown in yellow. Extracellular regions are indicated by the dashed line above the sequences while intracellular regions are indicated by the dashed line bellow the sequence (prediction obtained by <code><a href="../../reference/get_phobius.html">get_phobius()</a></code>).<br>
</li>
<li>Signal peptides (prediction obtained by <code><a href="../../reference/get_signalp.html">get_signalp()</a></code>) are indicated by the thick red line on the N-terminal side.<br>
</li>
<li>GPI attachment sites (as predicted by <code><a href="../../reference/get_big_pi.html">get_big_pi()</a></code> or <code><a href="../../reference/get_pred_gpi.html">get_pred_gpi()</a></code> depending on the <code><a href="../../reference/plot_prot.html">plot_prot()</a></code> argument <code>gpi</code>) are indicated by blue rhomboids.<br>
</li>
<li>Hydroxyprolines (as predicted by <code><a href="../../reference/predict_hyp.html">predict_hyp()</a></code>) are indicated by vertical dark gray lines.<br>
</li>
<li>Arabinogalactan glycomodule spans are indicated by the light grey background (motifs are found using <code><a href="../../reference/scan_ag.html">scan_ag()</a></code>).<br>
</li>
<li>Domains (prediction obtained by <code><a href="../../reference/get_hmm.html">get_hmm()</a></code>) are indicated by rectangles with an appropriate fill as indicated in the legend.</li>
</ul>
<p>Additionally by changing the argument <code>disorder</code> to <code>TRUE</code> disordered regions (prediction obtained by <code><a href="../../reference/get_espritz.html">get_espritz()</a></code>) are represented as zig-zag lines.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">p1 &lt;-<span class="st"> </span><span class="kw"><a href="../../reference/plot_prot.html">plot_prot</a></span>(<span class="dt">sequence =</span> seqs,
                <span class="dt">id =</span> ids,
                <span class="dt">disorder =</span> <span class="ot">TRUE</span>) <span class="co">#plot disordered regions</span>
p1</code></pre></div>
<div class="figure">
<img src="https://missuse.github.io/ragp/articles/PSV_p2.svg">
</div>
</div>
<div id="plot_prot-options" class="section level2">
<h2 class="hasAnchor">
<a href="#plot_prot-options" class="anchor"></a>plot_prot() options</h2>
<ul>
<li>
<strong>determining what features to plot</strong>: to turn off some elements from being plotted use the appropriate argument, for instance <code>hyp = FALSE</code> will result in hydroxyproline predictions to not be plotted (check full list of options: <code><a href="../../reference/plot_prot.html">?plot_prot</a></code>):</li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">p1 &lt;-<span class="st"> </span><span class="kw"><a href="../../reference/plot_prot.html">plot_prot</a></span>(<span class="dt">sequence =</span> seqs,
                <span class="dt">id =</span> ids,
                <span class="dt">disorder =</span> <span class="ot">TRUE</span>, <span class="co">#plot disordered regions</span>
                <span class="dt">hyp =</span> <span class="ot">FALSE</span>, <span class="co">#do not plot hydroxyprolines</span>
                <span class="dt">ag =</span> <span class="ot">FALSE</span>, <span class="co">#do not plot arabinogalactan motifs</span>
                <span class="dt">gpi =</span> <span class="st">"predgpi"</span>)  <span class="co">#use `get_pred_gpi()` to obtain GPI predictions</span>
p1</code></pre></div>
<div class="figure">
<img src="https://missuse.github.io/ragp/articles/PSV_p3.svg">
</div>
<ul>
<li>
<strong>feature specific options</strong>: passing appropriate function arguments to functions called by <code><a href="../../reference/plot_prot.html">plot_prot()</a></code> such as <code>dim</code> and <code>div</code> for <code><a href="../../reference/scan_ag.html">scan_ag()</a></code> or <code>tprob</code> for <code><a href="../../reference/predict_hyp.html">predict_hyp()</a></code> to change the rules of arabinogalactan motif scan and the threshold for decision if the proline is hydroxylated is supported:</li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">p2 &lt;-<span class="st"> </span><span class="kw"><a href="../../reference/plot_prot.html">plot_prot</a></span>(<span class="dt">sequence =</span> seqs,
                <span class="dt">id =</span> ids,
                <span class="dt">disorder =</span> <span class="ot">TRUE</span>, <span class="co">#plot disordered regions</span>
                <span class="dt">dim =</span> <span class="dv">5</span>, <span class="co">#`scan_ag()` argument, minimum number of dipeptides for a motif</span>
                <span class="dt">div =</span> <span class="dv">5</span>, <span class="co">#`scan_ag()` argument, maximum number of amino acids between dipeptides </span>
                <span class="dt">tprob =</span> <span class="fl">0.9</span>, <span class="co">#`predict_hyp()` argument, threshold probability for prediction</span>
                <span class="dt">model =</span> <span class="st">"Disprot"</span>) <span class="co">#use NMR model when calling `get_espritz()`</span>
p2</code></pre></div>
<div class="figure">
<img src="https://missuse.github.io/ragp/articles/PSV_p4.svg">
</div>
<ul>
<li>
<strong>change feature colors</strong>:</li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">p3 &lt;-<span class="st"> </span><span class="kw"><a href="../../reference/plot_prot.html">plot_prot</a></span>(<span class="dt">sequence =</span> seqs,
                <span class="dt">id =</span> ids,
                <span class="dt">hyp_col =</span> <span class="st">"chocolate2"</span>,
                <span class="dt">gpi_col =</span> <span class="st">"lightsalmon4"</span>,
                <span class="dt">nsp_col =</span> <span class="st">"sienna1"</span>,
                <span class="dt">ag_col =</span> <span class="st">"bisque2"</span>,
                <span class="dt">tm_col =</span> <span class="st">"aquamarine"</span>) 
p3</code></pre></div>
<div class="figure">
<img src="https://missuse.github.io/ragp/articles/PSV_p5.svg">
</div>
<ul>
<li>
<strong>change the order of domain plotting</strong>: in the above images it can be observed there are more entries in the legend then can be seen in the plot. Some domains cover other domains. By default the domains with the lowest I-evalue (independent e-value) are plotted on top. To change this the argument <code>dom_sort</code> can be used. This argument can take the following values: <code>"ievalue"</code> (default), <code>"abc"</code> (alphabetically sorted by domain names), <code>"cba"</code> (reverse of <code>"abc"</code>).</li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">p4 &lt;-<span class="st"> </span><span class="kw"><a href="../../reference/plot_prot.html">plot_prot</a></span>(<span class="dt">sequence =</span> seqs,
                <span class="dt">id =</span> ids,
                <span class="dt">dom_sort =</span> <span class="st">"cba"</span>)
p4</code></pre></div>
<div class="figure">
<img src="https://missuse.github.io/ragp/articles/PSV_dom_sort.svg">
</div>
<p>Ureide permease domain overlaps with TMEM144 transmembrane transporter domain and also protein kinase and protein tyrosine kinase domains overlap.</p>
</div>
<div id="customising-the-plot-using-ggplot-layers" class="section level2">
<h2 class="hasAnchor">
<a href="#customising-the-plot-using-ggplot-layers" class="anchor"></a>Customising the plot using ggplot layers</h2>
<p><code><a href="../../reference/plot_prot.html">plot_prot()</a></code> returns a <code>ggplot</code> object which can be further customized:</p>
<ul>
<li>
<strong>change theme elements</strong>:</li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">p1 <span class="op">+</span>
<span class="st">  </span><span class="kw">theme</span>(<span class="dt">legend.position =</span> <span class="st">"bottom"</span>,
        <span class="dt">legend.direction =</span> <span class="st">"vertical"</span>)</code></pre></div>
<div class="figure">
<img src="https://missuse.github.io/ragp/articles/PSV_p6.svg">
</div>
<ul>
<li>
<strong>add annotations</strong>:</li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">p1 <span class="op">+</span>
<span class="st">  </span><span class="kw">theme</span>(<span class="dt">legend.position =</span> <span class="st">"bottom"</span>,
        <span class="dt">legend.direction =</span> <span class="st">"vertical"</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">annotate</span>(<span class="dt">geom =</span> <span class="st">"text"</span>,
           <span class="dt">y =</span> <span class="dv">3</span>,
           <span class="dt">x =</span> <span class="dv">90</span>,
           <span class="dt">label =</span> <span class="st">"FAS domain"</span>)</code></pre></div>
<div class="figure">
<img src="https://missuse.github.io/ragp/articles/PSV_p7.svg">
</div>
<p>The <code>y</code> axis is continuous - <code>1</code> on the plot corresponds to Q9C7T7, <code>2</code> corresponds to Q9ZPR7 and 3 to Q9ZQ23. The <code>x</code> axis represents amino acids in the sequence. This enables adding annotations anywhere on the plot.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">p1 <span class="op">+</span>
<span class="st">  </span><span class="kw">theme</span>(<span class="dt">legend.position =</span> <span class="st">"bottom"</span>,
        <span class="dt">legend.direction =</span> <span class="st">"vertical"</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">annotate</span>(<span class="dt">geom =</span> <span class="st">"text"</span>,
           <span class="dt">y =</span> <span class="fl">1.6</span>,
           <span class="dt">x =</span> <span class="dv">530</span>,
           <span class="dt">label =</span> <span class="st">"annotation 1"</span>)</code></pre></div>
<div class="figure">
<img src="https://missuse.github.io/ragp/articles/PSV_p8.svg">
</div>
<ul>
<li>
<strong>change plot ratio</strong>: the ratio of the axes is determined automatically by <code><a href="../../reference/plot_prot.html">plot_prot()</a></code> and scales with the longest sequence plotted. To change this use the <code>ratio</code> argument to <code>coord_equal()</code>:</li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">p1 <span class="op">+</span>
<span class="st">  </span><span class="kw">theme</span>(<span class="dt">legend.position =</span> <span class="st">"bottom"</span>,
        <span class="dt">legend.direction =</span> <span class="st">"vertical"</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">coord_equal</span>(<span class="dt">ratio =</span> <span class="dv">80</span>,
              <span class="dt">expand =</span> <span class="ot">FALSE</span>)</code></pre></div>
<div class="figure">
<img src="https://missuse.github.io/ragp/articles/PSV_p9.svg">
</div>
<ul>
<li>
<strong>change domain colors</strong>: use another fill palette:</li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">p1 <span class="op">+</span>
<span class="st">  </span><span class="kw">theme</span>(<span class="dt">legend.position =</span> <span class="st">"bottom"</span>,
        <span class="dt">legend.direction =</span> <span class="st">"vertical"</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">scale_fill_brewer</span>(<span class="dt">type =</span> <span class="st">"div"</span>)</code></pre></div>
<div class="figure">
<img src="https://missuse.github.io/ragp/articles/PSV_p10.svg">
</div>
<p>or create a manual palette:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">p1 <span class="op">+</span>
<span class="st">  </span><span class="kw">theme</span>(<span class="dt">legend.position =</span> <span class="st">"bottom"</span>,
        <span class="dt">legend.direction =</span> <span class="st">"vertical"</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">scale_fill_manual</span>(<span class="dt">values =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"orange"</span>,
                               <span class="st">"red"</span>,
                               <span class="st">"yellow"</span>,
                               <span class="st">"dodgerblue"</span>,
                               <span class="st">"green"</span>,
                               <span class="st">"blue"</span>,
                               <span class="st">"purple"</span>,
                               <span class="st">"grey30"</span>))</code></pre></div>
<div class="figure">
<img src="https://missuse.github.io/ragp/articles/PSV_p11.svg">
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#basic-usage">Basic usage</a></li>
      <li><a href="#plot_prot-options">plot_prot() options</a></li>
      <li><a href="#customising-the-plot-using-ggplot-layers">Customising the plot using ggplot layers</a></li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Milan Dragicevic, Danijela Paunovic.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.0.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
