<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Getting started â€¢ ragp</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../../pkgdown.css" rel="stylesheet">
<script src="../../pkgdown.js"></script><meta property="og:title" content="Getting started">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-inverse navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../../index.html">ragp</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.3.0.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../../articles/pkgdown/start.html">Getting started</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../../articles/pkgdown/filter.html">Filtering hydroxyproline rich glycoproteins</a>
    </li>
    <li>
      <a href="../../articles/pkgdown/analyse.html">Analyse hydroxyproline rich glycoproteins</a>
    </li>
    <li>
      <a href="../../articles/pkgdown/PSV.html">Protein Sequence Visualization</a>
    </li>
  </ul>
</li>
<li>
  <a href="../../reference/index.html">Functions</a>
</li>
<li>
  <a href="../../news/index.html">News</a>
</li>
<li>
  <a href="https://ragp.shinyapps.io/Rapp/">ragp shiny</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/missuse/ragp/">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Getting started</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/missuse/ragp/blob/master/vignettes/pkgdown/start.Rmd"><code>vignettes/pkgdown/start.Rmd</code></a></small>
      <div class="hidden name"><code>start.Rmd</code></div>

    </div>

    
    
<p>This short section explains how get started using <code>ragp</code>, from installation and basic function arguments to manipulation of function outputs.</p>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p>There are several ways to install R packages hosted on git-hub, however the simplest is to use <code><a href="https://rdrr.io/pkg/devtools/man/reexports.html">devtools::install_github()</a></code> which will perform all the required steps automatically.</p>
<p>To install <code>ragp</code> run:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#install.packages("devtools") #if it is not installed on your system</span>
devtools<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/devtools/man/reexports.html">install_github</a></span>(<span class="st">"missuse/ragp"</span>)</code></pre></div>
<p>alternatively run:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># install.packages("devtools")</span>
devtools<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/devtools/man/reexports.html">install_github</a></span>(<span class="st">"missuse/ragp"</span>,
                         <span class="dt">build_vignettes =</span> <span class="ot">TRUE</span>)</code></pre></div>
<p>to build vignettes which can be viewed by:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/utils/browseVignettes.html">browseVignettes</a></span>(<span class="st">"ragp"</span>)</code></pre></div>
</div>
<div id="inputs" class="section level2">
<h2 class="hasAnchor">
<a href="#inputs" class="anchor"></a>Inputs</h2>
<p>Most <code>ragp</code> functions require single letter protein sequences and the corresponding identifiers as input. These can be provided in the form of basic <code>R</code> data types such as vectors or data frames. Additionally <code>ragp</code> imports the <a href="https://cran.r-project.org/web/packages/seqinr/index.html"><code>seqinr</code></a> package for the manipulation of <code>.FASTA</code> files, so the input objects can be a list of <code>SeqFastaAA</code> objects returned by the <code><a href="https://rdrr.io/pkg/seqinr/man/read.fasta.html">seqinr::read.fasta()</a></code>. The location of a <code>.FASTA</code> file is also possible as a type of input.</p>
<p>Input options will be illustrated using <code><a href="../../reference/scan_ag.html">scan_ag()</a></code> function:</p>
<ul>
<li>provide a <code>character</code> vector of protein sequences to the <code>sequence</code> argument and a <code>character</code> vector of protein identifiers to the <code>id</code> argument:</li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(ragp)
<span class="kw"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(at_nsp) <span class="co">#a data frame of 2700 Arabidopsis sequences</span>
input1 &lt;-<span class="st"> </span><span class="kw"><a href="../../reference/scan_ag.html">scan_ag</a></span>(<span class="dt">sequence =</span> at_nsp<span class="op">$</span>sequence,
                  <span class="dt">id =</span> at_nsp<span class="op">$</span>Transcript.id) </code></pre></div>
<ul>
<li>provide a <code>data.frame</code> to <code>data</code> argument, and names of columns containing the protein sequences and corresponding identifiers to <code>sequence</code> and <code>id</code> arguments:</li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">input2 &lt;-<span class="st"> </span><span class="kw"><a href="../../reference/scan_ag.html">scan_ag</a></span>(<span class="dt">data =</span> at_nsp,
                  <span class="dt">sequence =</span> <span class="st">"sequence"</span>,
                  <span class="dt">id =</span> <span class="st">"Transcript.id"</span>) </code></pre></div>
<p>quoting column names is not necessary:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">input3 &lt;-<span class="st"> </span><span class="kw"><a href="../../reference/scan_ag.html">scan_ag</a></span>(<span class="dt">data =</span> at_nsp,
                  <span class="dt">sequence =</span> sequence,
                  <span class="dt">id =</span> Transcript.id) </code></pre></div>
<ul>
<li>provide a list of <code>SeqFastaAA</code> objects to <code>data</code> argument:</li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(seqinr) <span class="co">#to create a fasta file with protein sequences</span>

<span class="co">#write a FASTA file</span>
seqinr<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/seqinr/man/write.fasta.html">write.fasta</a></span>(<span class="dt">sequence =</span> <span class="kw"><a href="https://rdrr.io/r/base/strsplit.html">strsplit</a></span>(at_nsp<span class="op">$</span>sequence, <span class="st">""</span>),
                    <span class="dt">name =</span> at_nsp<span class="op">$</span>Transcript.id, <span class="dt">file =</span> <span class="st">"at_nsp.fasta"</span>)

<span class="co">#read a FASTA file to a list of SeqFastaAA objects</span>
At_seq_fas &lt;-<span class="st"> </span><span class="kw">read.fasta</span>(<span class="st">"at_nsp.fasta"</span>,
                         <span class="dt">seqtype =</span>  <span class="st">"AA"</span>, 
                         <span class="dt">as.string =</span> <span class="ot">TRUE</span>) 

input4 &lt;-<span class="st"> </span><span class="kw"><a href="../../reference/scan_ag.html">scan_ag</a></span>(<span class="dt">data =</span> At_seq_fas) </code></pre></div>
<ul>
<li>provide the location of a <code>.FASTA</code> file to be analyzed as string:</li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">input5 &lt;-<span class="st"> </span><span class="kw"><a href="../../reference/scan_ag.html">scan_ag</a></span>(<span class="dt">data =</span> <span class="st">"at_nsp.fasta"</span>) <span class="co">#file at_nsp.fasta is in the working directory</span></code></pre></div>
<p>All of the outputs are equal:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/all.equal.html">all.equal</a></span>(input1,
          input2)
<span class="co">#&gt; [1] TRUE</span>

<span class="kw"><a href="https://rdrr.io/r/base/all.equal.html">all.equal</a></span>(input1,
          input3)
<span class="co">#&gt; [1] TRUE</span>

<span class="kw"><a href="https://rdrr.io/r/base/all.equal.html">all.equal</a></span>(input1,
          input4)
<span class="co">#&gt; [1] TRUE</span>

<span class="kw"><a href="https://rdrr.io/r/base/all.equal.html">all.equal</a></span>(input1,
          input5)
<span class="co">#&gt; [1] TRUE</span></code></pre></div>
<p>The only exceptions to this design are plotting functions <code><a href="../../reference/plot_prot.html">plot_prot()</a></code> and <code><a href="../../reference/plot_signalp.html">plot_signalp()</a></code> which require protein sequences to be supplied in the form of string vectors (input1 in the above example) and <code><a href="../../reference/pfam2go.html">pfam2go()</a></code> which does not take sequences as input.</p>
</div>
<div id="outputs" class="section level2">
<h2 class="hasAnchor">
<a href="#outputs" class="anchor"></a>Outputs</h2>
<p>All <code>ragp</code> functions return basic R data structures such as data frames, lists of vectors and lists of data frames, making them convenient for manipulation to anyone familiar with R. An especially effective way to manipulate these objects is by utilizing the <a href="https://www.tidyverse.org/"><code>tidyverse</code></a> collection of packages, especially <a href="https://dplyr.tidyverse.org/"><code>dplyr</code></a> and <a href="https://ggplot2.tidyverse.org/"><code>ggplot2</code></a>. Several <a href="https://dplyr.tidyverse.org/"><code>dplyr</code></a> functions that will be especially handy for data wrangling are:</p>
<ul>
<li><code><a href="https://dplyr.tidyverse.org/reference/join.html">dplyr::left_join()</a></code></li>
<li><code><a href="https://dplyr.tidyverse.org/reference/mutate.html">dplyr::mutate()</a></code></li>
<li><code><a href="https://dplyr.tidyverse.org/reference/group_by.html">dplyr::group_by()</a></code></li>
<li><code><a href="https://dplyr.tidyverse.org/reference/summarise.html">dplyr::summarise()</a></code></li>
<li><code><a href="https://dplyr.tidyverse.org/reference/filter.html">dplyr::filter()</a></code></li>
<li><code><a href="https://dplyr.tidyverse.org/reference/distinct.html">dplyr::distinct()</a></code></li>
</ul>
<p>Examples on usage of these functions on objects returned by <code>ragp</code> functions are provided in <a href="https://missuse.github.io/ragp/articles/pkgdown/filter">HRGP filtering</a> and <a href="https://missuse.github.io/ragp/articles/pkgdown/analyse">HRGP analysis</a> tutorials. Additionally there are extensive examples on the internet on usage of the mentioned functions.</p>
<p>Obtaining pretty visualizations is usually the goal of the above mentioned data manipulations. The golden standard of R graphics at present is the <a href="https://ggplot2.tidyverse.org/">ggplot2</a> package and we recommend it to graphically summarize the data. Additionally <code>ragp</code> contains <code><a href="../../reference/plot_prot.html">plot_prot()</a></code> function which is a wrapper for <code>ggplot2</code>, and while <code><a href="../../reference/plot_prot.html">plot_prot()</a></code> can be used without knowing <code>ggplot2</code> syntax, to tweak the plot style at least a basic knowledge of <code>ggplot2</code> is required. Examples are provided in <a href="https://missuse.github.io/ragp/articles/pkgdown/PSV">protein sequence visualization</a> tutorial.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#installation">Installation</a></li>
      <li><a href="#inputs">Inputs</a></li>
      <li><a href="#outputs">Outputs</a></li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Milan Dragicevic, Danijela Paunovic.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.0.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
