<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Mining for Hydroxyproline rich glycoprotein sequences • ragp</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Mining for Hydroxyproline rich glycoprotein sequences">
<meta property="og:description" content="Identification of plant hydroxyproline rich glycoprotein sequences
    by performing sequence bias and motif search, prediction of
    hydroxylated proline residues and scraping relevant data from
    web based prediction services.">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">ragp</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.2.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/ragp-vignette.html">ragp vignette</a>
    </li>
  </ul>
</li>
<li>
  <a href="news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/missuse/ragp">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    

    
    
<!-- README.md is generated from README.Rmd. Please edit that file -->
<div id="ragp" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#ragp" class="anchor"></a>ragp</h1></div>

<p><code>ragp</code> is primarily designed for mining and analysis of plant Hydroxyproline rich glycoproteins. It incorporates a novel concept with an additional analysis layer where the probability of proline hydroxylation is estimated by a machine learning model. <code>ragp</code> can also be used for protein annotation by obtaining predictions for several protein features based on sequence (secretory signals, transmembrane regions, domains, glycosylphosphatidylinositol attachment sites and disordered regions). Additionally ragp provides tools for visualization of the mentioned attributes.</p>
<p>Short example:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(ragp)
ids &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"Q9FLL2"</span>, <span class="co">#several uniprot accessions</span>
         <span class="st">"Q9LS14"</span>,
         <span class="st">"Q9S7I8"</span>,
         <span class="st">"Q9M2Z2"</span>,
         <span class="st">"Q9FIN5"</span>)

seqs &lt;-<span class="st"> </span><span class="kw">unlist</span>(protr<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/protr/topics/getUniProt">getUniProt</a></span>(ids)) <span class="co">#download sequences </span>

p1 &lt;-<span class="st"> </span><span class="kw"><a href="reference/plot_prot.html">plot_prot</a></span>(seqs, <span class="co">#plot sequence features</span>
                ids,
                <span class="dt">hyp =</span> <span class="ot">FALSE</span>, <span class="co">#do not plot hydroxyprolines</span>
                <span class="dt">ag =</span> <span class="ot">FALSE</span>) <span class="co">#do not plot ag spans</span>
p1</code></pre></div>
<div class="figure">
<img src="https://missuse.github.io/ragp/reference/figures/README-plot_prot-2.svg" alt="alt text"><p class="caption">alt text</p>
</div>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p>You can install ragp from github with:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># install.packages("devtools")</span>
devtools<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/devtools/topics/install_github">install_github</a></span>(<span class="st">"missuse/ragp"</span>)</code></pre></div>
<p>Or alternatively</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># install.packages("devtools")</span>
devtools<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/devtools/topics/install_github">install_github</a></span>(<span class="st">"missuse/ragp"</span>,
                         <span class="dt">build_vignettes =</span> <span class="ot">TRUE</span>)</code></pre></div>
<p>to build vignettes which can be viewed by:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">browseVignettes</span>(<span class="st">"ragp"</span>)</code></pre></div>
</div>
<div id="usage" class="section level2">
<h2 class="hasAnchor">
<a href="#usage" class="anchor"></a>Usage</h2>
<p><code>ragp</code> functions require single letter protein sequences and the corresponding identifiers as input. These can be provided in the form of basic <code>R</code> data types such as vectors or data frames. Additionally <code>ragp</code> imports the <a href="https://cran.r-project.org/web/packages/seqinr/index.html"><code>seqinr</code></a> package for the manipulation of <code>.FASTA</code> files, so the input objects can be a list of <code>SeqFastaAA</code> objects returned by the <a href="https://cran.r-project.org/web/packages/seqinr/index.html"><code>seqinr</code></a> function <code>read.fasta</code>. The location of a <code>.FASTA</code> file is also possible as a type of input.</p>
</div>
<div id="input" class="section level2">
<h2 class="hasAnchor">
<a href="#input" class="anchor"></a>Input</h2>
<p>Input options will be illustrated using <code><a href="reference/scan_ag.html">scan_ag()</a></code> function:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(ragp)
<span class="kw">data</span>(at_nsp) <span class="co">#a data frame of 2700 Arabidopsis sequences</span>
input1 &lt;-<span class="st"> </span><span class="kw"><a href="reference/scan_ag.html">scan_ag</a></span>(<span class="dt">sequence =</span> at_nsp<span class="op">$</span>sequence,
                  <span class="dt">id =</span> at_nsp<span class="op">$</span>Transcript.id) </code></pre></div>
<p>equivalent to:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">input2 &lt;-<span class="st"> </span><span class="kw"><a href="reference/scan_ag.html">scan_ag</a></span>(<span class="dt">data =</span> at_nsp,
                  <span class="dt">sequence =</span> <span class="st">"sequence"</span>,
                  <span class="dt">id =</span> <span class="st">"Transcript.id"</span>) </code></pre></div>
<p>same as without quotation:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">input3 &lt;-<span class="st"> </span><span class="kw"><a href="reference/scan_ag.html">scan_ag</a></span>(<span class="dt">data =</span> at_nsp,
                  <span class="dt">sequence =</span> sequence,
                  <span class="dt">id =</span> Transcript.id) </code></pre></div>
<p>A list of <code>SeqFastaAA</code> objects is also a viable input:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(seqinr) <span class="co">#to create a fasta file with protein sequences</span>

<span class="co">#write a FASTA file</span>
seqinr<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/seqinr/topics/write.fasta">write.fasta</a></span>(<span class="dt">sequence =</span> <span class="kw">strsplit</span>(at_nsp<span class="op">$</span>sequence, <span class="st">""</span>),
                    <span class="dt">name =</span> at_nsp<span class="op">$</span>Transcript.id, <span class="dt">file =</span> <span class="st">"at_nsp.fasta"</span>)

<span class="co">#read a FASTA file</span>
At_seq_fas &lt;-<span class="st"> </span><span class="kw">read.fasta</span>(<span class="st">"at_nsp.fasta"</span>,
                         <span class="dt">seqtype =</span>  <span class="st">"AA"</span>, 
                         <span class="dt">as.string =</span> <span class="ot">TRUE</span>) 

input4 &lt;-<span class="st"> </span><span class="kw"><a href="reference/scan_ag.html">scan_ag</a></span>(<span class="dt">data =</span> At_seq_fas) </code></pre></div>
<p>and lastly the location of a <code>.FASTA</code> file to be analyzed as string:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">input5 &lt;-<span class="st"> </span><span class="kw"><a href="reference/scan_ag.html">scan_ag</a></span>(<span class="dt">data =</span> <span class="st">"at_nsp.fasta"</span>) </code></pre></div>
</div>
<div id="n-sp-query" class="section level2">
<h2 class="hasAnchor">
<a href="#n-sp-query" class="anchor"></a>N-sp query</h2>
<p>HRGPs are secreted proteins, therefore they are expected to contain a secretory signal sequence on the N-terminus (N-sp). <code>ragp</code> incorporates N-sp prediction by querying <a href="http://www.cbs.dtu.dk/services/SignalP/">SignalP</a>, <a href="http://www.cbs.dtu.dk/services/TargetP/">TargetP</a> <span class="citation">(Emanuelsson et al. 2000)</span> and <a href="http://phobius.sbc.su.se/">Phobius</a> <span class="citation">(Käll, Krogh, and Sonnhammer 2007)</span> in an efficient manner via the functions: <code><a href="reference/get_signalp.html">get_signalp()</a></code>, <code><a href="reference/get_targetp.html">get_targetp()</a></code> and <code><a href="reference/get_phobius.html">get_phobius()</a></code>.</p>
<p>To query <a href="http://www.cbs.dtu.dk/services/SignalP/">SignalP</a> predictions:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">nsp_signalp &lt;-<span class="st"> </span><span class="kw"><a href="reference/get_signalp.html">get_signalp</a></span>(at_nsp,
                           sequence,
                           Transcript.id)</code></pre></div>
<p>The predictions for the 2700 sequences contained in <code>at_nsp</code> data frame should be available in around 1 minute. The returned object <code>nsp_signalp</code> is a data frame resembling the web servers output:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">knitr<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/knitr/topics/kable">kable</a></span>(<span class="kw">head</span>(nsp_signalp))</code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">id</th>
<th align="left">Cmax</th>
<th align="left">Cmax.pos</th>
<th align="left">Ymax</th>
<th align="left">Ymax.pos</th>
<th align="left">Smax</th>
<th align="left">Smax.pos</th>
<th align="left">Smean</th>
<th align="left">Dmean</th>
<th align="left">is.sp</th>
<th align="left">Dmaxcut</th>
<th align="left">Networks.used</th>
<th align="left">is.signalp</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">ATCG00660.1</td>
<td align="left">0.210</td>
<td align="left">30</td>
<td align="left">0.162</td>
<td align="left">30</td>
<td align="left">0.260</td>
<td align="left">2</td>
<td align="left">0.146</td>
<td align="left">0.154</td>
<td align="left">N</td>
<td align="left">0.450</td>
<td align="left">SignalP-noTM</td>
<td align="left">FALSE</td>
</tr>
<tr class="even">
<td align="left">AT2G43600.1</td>
<td align="left">0.860</td>
<td align="left">23</td>
<td align="left">0.894</td>
<td align="left">23</td>
<td align="left">0.984</td>
<td align="left">13</td>
<td align="left">0.930</td>
<td align="left">0.913</td>
<td align="left">Y</td>
<td align="left">0.450</td>
<td align="left">SignalP-noTM</td>
<td align="left">TRUE</td>
</tr>
<tr class="odd">
<td align="left">AT2G28410.1</td>
<td align="left">0.779</td>
<td align="left">23</td>
<td align="left">0.826</td>
<td align="left">23</td>
<td align="left">0.940</td>
<td align="left">15</td>
<td align="left">0.877</td>
<td align="left">0.853</td>
<td align="left">Y</td>
<td align="left">0.450</td>
<td align="left">SignalP-noTM</td>
<td align="left">TRUE</td>
</tr>
<tr class="even">
<td align="left">AT2G22960.1</td>
<td align="left">0.701</td>
<td align="left">23</td>
<td align="left">0.790</td>
<td align="left">23</td>
<td align="left">0.948</td>
<td align="left">15</td>
<td align="left">0.891</td>
<td align="left">0.844</td>
<td align="left">Y</td>
<td align="left">0.450</td>
<td align="left">SignalP-noTM</td>
<td align="left">TRUE</td>
</tr>
<tr class="odd">
<td align="left">AT2G19580.1</td>
<td align="left">0.422</td>
<td align="left">26</td>
<td align="left">0.586</td>
<td align="left">26</td>
<td align="left">0.885</td>
<td align="left">17</td>
<td align="left">0.798</td>
<td align="left">0.671</td>
<td align="left">Y</td>
<td align="left">0.500</td>
<td align="left">SignalP-TM</td>
<td align="left">TRUE</td>
</tr>
<tr class="even">
<td align="left">AT2G19690.2</td>
<td align="left">0.797</td>
<td align="left">29</td>
<td align="left">0.870</td>
<td align="left">29</td>
<td align="left">0.987</td>
<td align="left">18</td>
<td align="left">0.952</td>
<td align="left">0.914</td>
<td align="left">Y</td>
<td align="left">0.450</td>
<td align="left">SignalP-noTM</td>
<td align="left">TRUE</td>
</tr>
</tbody>
</table>
<p>Similarly <a href="http://phobius.sbc.su.se/">Phobius</a> and <a href="http://www.cbs.dtu.dk/services/TargetP/">TargetP</a> can be accessed by the functions <code><a href="reference/get_phobius.html">get_phobius()</a></code> and <code><a href="reference/get_targetp.html">get_targetp()</a></code>.</p>
<p>When a more in depth look at SignalP predictions is needed the function <code>plot_signalp</code> can be used. This function accepts a single protein sequence as input:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pred &lt;-<span class="st"> </span><span class="kw"><a href="reference/plot_signalp.html">plot_signalp</a></span>(<span class="dt">sequence =</span> at_nsp<span class="op">$</span>sequence[<span class="dv">5</span>],
                     <span class="dt">id =</span> at_nsp<span class="op">$</span>Transcript.id[<span class="dv">5</span>])</code></pre></div>
<div class="figure">
<img src="https://missuse.github.io/ragp/reference/figures/README-SignalP3-1.png" alt="alt text"><p class="caption">alt text</p>
</div>
</div>
<div id="gpi-and-hmm-query" class="section level2">
<h2 class="hasAnchor">
<a href="#gpi-and-hmm-query" class="anchor"></a>GPI and hmm query</h2>
<p>HRGPs, and especially AGPs are often linked to the membrane by a glycosylphosphatidylinositol (GPI) anchor <span class="citation">(Ellis et al. 2010)</span>. To fetch the GPI modification site predictions from <a href="http://mendel.imp.ac.at/gpi/plant_server.html">big-PI</a> <span class="citation">(B. Eisenhaber et al. 2003)</span> the function <code><a href="reference/get_big_pi.html">get_big_pi()</a></code> can be used:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">big_pi_pred &lt;-<span class="st"> </span><span class="kw"><a href="reference/get_big_pi.html">get_big_pi</a></span>(<span class="dt">data =</span> at_nsp[<span class="dv">1</span><span class="op">:</span><span class="dv">100</span>,],
                          <span class="dt">sequence =</span> sequence,
                          <span class="dt">id =</span> Transcript.id)</code></pre></div>
<p>To use <a href="http://gpcr.biocomp.unibo.it/predgpi/">PredGPI</a> <span class="citation">(Pierleoni, Martelli, and Casadio 2008)</span> for GPI prediction:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pred_gpi_pred &lt;-<span class="st"> </span><span class="kw"><a href="reference/get_pred_gpi.html">get_pred_gpi</a></span>(<span class="dt">data =</span> at_nsp[<span class="dv">1</span><span class="op">:</span><span class="dv">100</span>,],
                              <span class="dt">sequence =</span> sequence,
                              <span class="dt">id =</span> Transcript.id)</code></pre></div>
<p>Similarly, domains can be identified by <a href="https://www.ebi.ac.uk/Tools/hmmer/search/hmmscan">hmmscan</a> <span class="citation">(Finn, Clements, and Eddy 2011)</span> using <code><a href="reference/get_hmm.html">get_hmm()</a></code> function:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pfam_pred &lt;-<span class="st"> </span><span class="kw"><a href="reference/get_hmm.html">get_hmm</a></span>(<span class="dt">data =</span> at_nsp[<span class="dv">1</span><span class="op">:</span><span class="dv">20</span>,],
                     <span class="dt">sequence =</span> sequence,
                     <span class="dt">id =</span> Transcript.id,
                     <span class="dt">verbose =</span> <span class="ot">FALSE</span>)</code></pre></div>
<p>To enrich with GO terms:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pfam_pred_go &lt;-<span class="st"> </span><span class="kw"><a href="reference/pfam2go.html">pfam2go</a></span>(pfam_pred)</code></pre></div>
<p>first 10 rows of the GO result with selected columns:</p>
<table class="table">
<thead><tr class="header">
<th align="left">id</th>
<th align="left">name</th>
<th align="left">acc</th>
<th align="left">GO_name</th>
<th align="left">GO_acc</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">ATCG00660.1</td>
<td align="left">Ribosomal_L20</td>
<td align="left">PF00453.18</td>
<td align="left"><a href="GO:translation" class="uri">GO:translation</a></td>
<td align="left"><a href="GO:0006412" class="uri">GO:0006412</a></td>
</tr>
<tr class="even">
<td align="left">ATCG00660.1</td>
<td align="left">Ribosomal_L20</td>
<td align="left">PF00453.18</td>
<td align="left"><a href="GO:ribosome" class="uri">GO:ribosome</a></td>
<td align="left"><a href="GO:0005840" class="uri">GO:0005840</a></td>
</tr>
<tr class="odd">
<td align="left">ATCG00660.1</td>
<td align="left">Ribosomal_L20</td>
<td align="left">PF00453.18</td>
<td align="left"><a href="GO:intracellular" class="uri">GO:intracellular</a></td>
<td align="left"><a href="GO:0005622" class="uri">GO:0005622</a></td>
</tr>
<tr class="even">
<td align="left">ATCG00660.1</td>
<td align="left">Ribosomal_L20</td>
<td align="left">PF00453.18</td>
<td align="left">
<a href="GO:structural" class="uri">GO:structural</a> constituent of ribosome</td>
<td align="left"><a href="GO:0003735" class="uri">GO:0003735</a></td>
</tr>
<tr class="odd">
<td align="left">ATCG00660.1</td>
<td align="left">Ribosomal_L20</td>
<td align="left">PF00453.18</td>
<td align="left">
<a href="GO:rRNA" class="uri">GO:rRNA</a> binding</td>
<td align="left"><a href="GO:0019843" class="uri">GO:0019843</a></td>
</tr>
<tr class="even">
<td align="left">AT2G43600.1</td>
<td align="left">Glyco_hydro_19</td>
<td align="left">PF00182.19</td>
<td align="left">
<a href="GO:cell" class="uri">GO:cell</a> wall macromolecule catabolic process</td>
<td align="left"><a href="GO:0016998" class="uri">GO:0016998</a></td>
</tr>
<tr class="odd">
<td align="left">AT2G43600.1</td>
<td align="left">Glyco_hydro_19</td>
<td align="left">PF00182.19</td>
<td align="left">
<a href="GO:chitin" class="uri">GO:chitin</a> catabolic process</td>
<td align="left"><a href="GO:0006032" class="uri">GO:0006032</a></td>
</tr>
<tr class="even">
<td align="left">AT2G43600.1</td>
<td align="left">Glyco_hydro_19</td>
<td align="left">PF00182.19</td>
<td align="left">
<a href="GO:chitinase" class="uri">GO:chitinase</a> activity</td>
<td align="left"><a href="GO:0004568" class="uri">GO:0004568</a></td>
</tr>
<tr class="odd">
<td align="left">AT2G43600.1</td>
<td align="left">Glyco_hydro_19</td>
<td align="left">PF00182.19</td>
<td align="left">
<a href="GO:cell" class="uri">GO:cell</a> wall macromolecule catabolic process</td>
<td align="left"><a href="GO:0016998" class="uri">GO:0016998</a></td>
</tr>
<tr class="even">
<td align="left">AT2G43600.1</td>
<td align="left">Glyco_hydro_19</td>
<td align="left">PF00182.19</td>
<td align="left">
<a href="GO:chitin" class="uri">GO:chitin</a> catabolic process</td>
<td align="left"><a href="GO:0006032" class="uri">GO:0006032</a></td>
</tr>
</tbody>
</table>
</div>
<div id="disorder-prediction" class="section level2">
<h2 class="hasAnchor">
<a href="#disorder-prediction" class="anchor"></a>Disorder prediction</h2>
<p>HRGPs are considered to be intrinsically disordered. To identify potential disordered regions in proteins <code>ragp</code> contains <code><a href="reference/get_espritz.html">get_espritz()</a></code> function which queries <a href="http://protein.bio.unipd.it/espritz/">ESpritz</a> <span class="citation">(Walsh et al. 2012)</span> web server:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">at_espritz &lt;-<span class="st"> </span><span class="kw"><a href="reference/get_espritz.html">get_espritz</a></span>(at_nsp[<span class="dv">1</span><span class="op">:</span><span class="dv">100</span>,],
                          sequence,
                          Transcript.id)</code></pre></div>
</div>
<div id="maab-classification" class="section level2">
<h2 class="hasAnchor">
<a href="#maab-classification" class="anchor"></a>MAAB classification</h2>
<p>The MAAB pipeline is explained in detail in <span class="citation">Johnson et al. (2017)</span>. The <code>ragp</code> function <code><a href="reference/maab.html">maab()</a></code> performs motif and amino acid bias classification of HRGPs:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">maab_class &lt;-<span class="st"> </span><span class="kw"><a href="reference/maab.html">maab</a></span>(at_nsp,
                   sequence,
                   Transcript.id) 
                 </code></pre></div>
<p>MAAB classification relies on knowledge if the protein is bound to the membrane by a GPI. One way to remove ambiguities in classes is to set <code>get_gpi</code> argument to <code>"bigpi"</code> or <code>"predgpi"</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">maab_class &lt;-<span class="st"> </span><span class="kw"><a href="reference/maab.html">maab</a></span>(at_nsp,
                   sequence,
                   Transcript.id,
                   <span class="dt">get_gpi =</span> <span class="st">"bigpi"</span>)</code></pre></div>
<p>This will run <code><a href="reference/get_big_pi.html">get_big_pi()</a></code> (or <code><a href="reference/get_pred_gpi.html">get_pred_gpi()</a></code> if <code>"predgpi"</code> was set) only on the sequences that belong to one of the HRGP classes.</p>
</div>
<div id="hydroxyproline-prediction" class="section level2">
<h2 class="hasAnchor">
<a href="#hydroxyproline-prediction" class="anchor"></a>hydroxyproline prediction</h2>
<p>The key feature of HRGPs is the presence of hydroxyprolines (Hyp, O) which represent glycosylation sites. While many HRGPs can be found based on biased amino acid composition, or the presence of certain amino acid motifs, there exists an abundance of chimeric proteins comprised from specific domains and HRGP motifs which are much harder to identify based on the mentioned features. In an attempt to identify these types of sequences <code>ragp</code> incorporates a model specifically built to predict the probability of proline hydroxylation in plant proteins:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">hyp_pred &lt;-<span class="st"> </span><span class="kw"><a href="reference/predict_hyp.html">predict_hyp</a></span>(at_nsp,
                        sequence,
                        Transcript.id)</code></pre></div>
<p>Output is a list of two elements. The first element <code>prediction</code> is a data frame. First 10 rows:</p>
<table class="table">
<thead><tr class="header">
<th align="left">id</th>
<th align="left">substr</th>
<th align="right">P_pos</th>
<th align="right">prob</th>
<th align="left">HYP</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">AT2G43600.1</td>
<td align="left">FSQNCMDTSCPGLKECCSRWG</td>
<td align="right">31</td>
<td align="right">0.02</td>
<td align="left">No</td>
</tr>
<tr class="even">
<td align="left">AT2G43600.1</td>
<td align="left">EYCGFFCFSGPCNIKGKSYGY</td>
<td align="right">58</td>
<td align="right">0.02</td>
<td align="left">No</td>
</tr>
<tr class="odd">
<td align="left">AT2G43600.1</td>
<td align="left">YGYDYNVDAGPRGKIETVITS</td>
<td align="right">76</td>
<td align="right">0.03</td>
<td align="left">No</td>
</tr>
<tr class="even">
<td align="left">AT2G43600.1</td>
<td align="left">ERYCSKSKKYPCEPGKNYYGR</td>
<td align="right">163</td>
<td align="right">0.04</td>
<td align="left">No</td>
</tr>
<tr class="odd">
<td align="left">AT2G43600.1</td>
<td align="left">CSKSKKYPCEPGKNYYGRGLL</td>
<td align="right">166</td>
<td align="right">0.02</td>
<td align="left">No</td>
</tr>
<tr class="even">
<td align="left">AT2G43600.1</td>
<td align="left">YYGAGKHLGLPLLKDPDLVSR</td>
<td align="right">194</td>
<td align="right">0.02</td>
<td align="left">No</td>
</tr>
<tr class="odd">
<td align="left">AT2G43600.1</td>
<td align="left">KHLGLPLLKDPDLVSRSPEVA</td>
<td align="right">199</td>
<td align="right">0.02</td>
<td align="left">No</td>
</tr>
<tr class="even">
<td align="left">AT2G43600.1</td>
<td align="left">LKDPDLVSRSPEVAFKFAMWF</td>
<td align="right">206</td>
<td align="right">0.02</td>
<td align="left">No</td>
</tr>
<tr class="odd">
<td align="left">AT2G43600.1</td>
<td align="left">AMWFWNRNVRPALYLGFGEIT</td>
<td align="right">223</td>
<td align="right">0.02</td>
<td align="left">No</td>
</tr>
<tr class="even">
<td align="left">AT2G43600.1</td>
<td align="left">EMLGVTPDQGLDC</td>
<td align="right">267</td>
<td align="right">0.00</td>
<td align="left">No</td>
</tr>
</tbody>
</table>
<p><code><a href="reference/predict_hyp.html">predict_hyp()</a></code> is also available as a <a href="https://profenicolalac.shinyapps.io/HYPpredict_Shiny/">shiny app</a>. Details on how hydroxyproline sites are predicted will be available soon.</p>
</div>
<div id="scan-ag-glycomodules" class="section level2">
<h2 class="hasAnchor">
<a href="#scan-ag-glycomodules" class="anchor"></a>scan AG glycomodules</h2>
<p>AGPs are characterized by the presence of so called AG glycomodules - amino acid dipeptides: OA, OS, OT, AO, SO and TO (and probably OG, OV, GO and VO) which are in close proximity to each other. Where: O - hydroxyproline, A - alanine, S - serine, T - threnonine, G - glycine and V - valine. <code><a href="reference/scan_ag.html">scan_ag()</a></code> function attempts to find the mentioned dipeptides according to user specified rules. Example:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">at_nsp_ag &lt;-<span class="st"> </span><span class="kw"><a href="reference/scan_ag.html">scan_ag</a></span>(at_nsp,
                     sequence,
                     Transcript.id,
                     <span class="dt">dim =</span> <span class="dv">3</span>, <span class="co">#at least 3 dimers must be present</span>
                     <span class="dt">div =</span> <span class="dv">10</span>, <span class="co">#no more than 10 amino acids apart</span>
                     <span class="dt">type =</span> <span class="st">"conservative"</span>) <span class="co">#dimers will be defined as: PA, PS, PT, AP, SP, TP</span></code></pre></div>
<p>List several sequences. possible AG glycomodules are in caps:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ind &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dv">145</span>, <span class="dv">147</span>, <span class="dv">160</span>, <span class="dv">170</span>,
         <span class="dv">189</span>, <span class="dv">203</span>, <span class="dv">214</span>, <span class="dv">224</span>)
at_nsp_ag<span class="op">$</span>sequence[ind]
<span class="co">#&gt; [1] "mayatilmifsvvalmsgerahaavdcsslilnmadclsfvtsgstvvkpegtccsglktvvrtgpeclceafknsgslgltldlskaaslPSvckvaAPPSarcglsvsgdpPAtAPglSPTagagAPAlssganaaTPvsSPrssdasllsvsfafvifmalissfy"                                                                                                                                                                                                       </span>
<span class="co">#&gt; [2] "mgyrrsyaitfvalvaalwsvtkaqPSsscvstlttlSPclsyitgnstTPSqpccsrldsviksspqcicsavnspipniglninrtqalqlpnacniqTPpltqcnaatgPTaqpPAPSPTekTPdvtlTPTslpgarsgvgggsktvpsvgtgsssrnvdplplhflmfavlvvctssfl"                                                                                                                                                                                         </span>
<span class="co">#&gt; [3] "mrlllsllfllalttyssatyclcrdgvgekdlqtsidyacgvlkdcnpihekgpcyqpntikshcdwavntyfqrfgqisgscnfsgtattsqnlPStvvtgclyPSSPgsagtTPTtgTPSgtqtfpgpPAfgPAgdfdPSgnngAPSlfisialslgfsvviafl"                                                                                                                                                                                                        </span>
<span class="co">#&gt; [4] "mhhhhhpcnrkpfttifsffllylnlhnqqiiearnpsqfttnpspdvsipeikrhlqqygylpqnkesddvsfeqalvryqknlglpitgkpdsdtlsqillprcgfpddvepktapfhtgkkyvyfpgrprwtrdvplkltyafsqenltpylaptdirrvfrrafgkwasvipvsfietedyviadikigffngdhgdgepfdgvlgvlahtfspengrlhldkaetwavdfdeekssvavdlesvavheighvlglghssvkdaamyptlkprskkvnlnmddvvgvqslygtnpnftlnsllasetstnladgsrirsqgmiystlstvialcflnw"                          </span>
<span class="co">#&gt; [5] "makmqlsifiavvalivcsasaktaSPPAPvlpPTPAPAPAPenvnltellsvagpfhtfldyllstgvietfqnqannteegitifvpkddafkaqknpplsnltkdqlkqlvlfhalphyyslsefknlsqsgpvstfaggqyslkftdvsgtvridslwtrtkvsssvfstdpvavyqvnrvllpeaifgtdvppmPAPAPAPivsAPSdSPSvadsegasSPksshknsgqklllapismvisglvalfl"                                                                                                                  </span>
<span class="co">#&gt; [6] "mmrgaaptgvvsvmvlmilvllkqiesasangsslglpprkfcniyqgswvydksyplydskncpfierqfncksngrpdseylkyrwqpsgcnlprfngldflgrimkgkklmfvgdslslnqwqsltcllhnaAPkanststrSPSglsvfsfPAynssimfsrnaflvdivgappkrvmkldsissgslwktadvlvfnswhwwlhtdrkqpwdaimsgnvtvkdmdrlvayekammtwakwidqnidpsktkvffqgispdhgrarewskqggkgscigetkpimgssylagphaaemvvakviktmknqarlmdvtlmsqlrkdghpsvygfgghrmadcshwclsgvpdswnqllyselfhs"</span>
<span class="co">#&gt; [7] "maliknnifftsllifvtlfgvavggtvhkvgntkgwtmiggdyeawassrvfqvgdtlvfaynkdyhdvtevthndfemcesskplrryktgsdsisltkpglqhficgvpghckkgqklqihvlpaslghvavpvpgpvrsqssssSPSPSPlvdppvnnAPqyqmgPTPAshsaasadfiftfsfdltlidlctffilffilv"                                                                                                                                                                  </span>
<span class="co">#&gt; [8] "marsfaiavicivliagvtgqAPTSPPTaTPAPPTPTTPpPAaTPppvsAPppvttSPppvttAPpPAnppppvsSPpPASPpPATPppvaSPpppvaSPpPATPppvaTPpPAPlaSPPAqvPAPAPTtkpdSPSPSPSsSPplPSsdAPgPStdsiSPAPSPTdvndqngaskmvsslvfgsvlvwfmi"</span></code></pre></div>
<p>Before I wrote OA, OS, OT, AO, SO and TO (and probably OG, OV, GO and VO) are AG glycomodules but the above output considers P instead of O since most of the time the positions of O are unknown. However if the sequence argument to <code><a href="reference/scan_ag.html">scan_ag()</a></code> contains O’s at some positions, <code><a href="reference/scan_ag.html">scan_ag()</a></code> will consider only them. To do this the sequence output of <code><a href="reference/predict_hyp.html">predict_hyp()</a></code> function can be used. Example:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">at_nsp_ag &lt;-<span class="st"> </span><span class="kw"><a href="reference/scan_ag.html">scan_ag</a></span>(<span class="dt">data =</span> hyp_pred<span class="op">$</span>sequence, <span class="co">#hyp_pred was created a little bit back</span>
                     <span class="dt">sequence =</span> sequence, 
                     <span class="dt">id =</span> id,
                     <span class="dt">dim =</span> <span class="dv">3</span>, <span class="co">#at least 3 dimers must be present</span>
                     <span class="dt">div =</span> <span class="dv">10</span>, <span class="co">#no more than 10 amino acids apart</span>
                     <span class="dt">type =</span> <span class="st">"conservative"</span>)
<span class="co">#&gt; sequence vector contains O, O will be considered instead of P</span>

at_nsp_ag<span class="op">$</span>sequence[ind]
<span class="co">#&gt; [1] "mayatilmifsvvalmsgerahaavdcsslilnmadclsfvtsgstvvkpegtccsglktvvrtgpeclceafknsgslgltldlskaaslpsvckvaappsarcglsvsgdpOAtAOglSOTagagAOAlssganaatpvssprssdasllsvsfafvifmalissfy"                                                                                                                                                                                                       </span>
<span class="co">#&gt; [2] "mgyrrsyaitfvalvaalwsvtkaqpssscvstlttlspclsyitgnsttpsqpccsrldsviksspqcicsavnspipniglninrtqalqlpnacniqtppltqcnaatgOTaqpOAOSOTekTOdvtlTOTslogarsgvgggsktvpsvgtgsssrnvdplplhflmfavlvvctssfl"                                                                                                                                                                                         </span>
<span class="co">#&gt; [3] "mrlllsllfllalttyssatyclcrdgvgekdlqtsidyacgvlkdcnpihekgpcyqpntikshcdwavntyfqrfgqisgscnfsgtattsqnlpstvvtgclypsSOgsagtTOTtgTOSgtqtfpgoOAfgOAgdfdpsgnngapslfisialslgfsvviafl"                                                                                                                                                                                                        </span>
<span class="co">#&gt; [4] "mhhhhhpcnrkpfttifsffllylnlhnqqiiearnpsqfttnpsodvsipeikrhlqqygylpqnkesddvsfeqalvryqknlglpitgkpdsdtlsqillprcgfpddvepktapfhtgkkyvyfpgrprwtrdvplkltyafsqenltpylaptdirrvfrrafgkwasvipvsfietedyviadikigffngdhgdgepfdgvlgvlahtfspengrlhldkaetwavdfdeekssvavdlesvavheighvlglghssvkdaamyptlkprskkvnlnmddvvgvqslygtnpnftlnsllasetstnladgsrirsqgmiystlstvialcflnw"                          </span>
<span class="co">#&gt; [5] "makmqlsifiavvalivcsasaktaSOOAOvloOTOAOAOAOenvnltellsvagpfhtfldyllstgvietfqnqannteegitifvpkddafkaqknpplsnltkdqlkqlvlfhalphyyslsefknlsqsgpvstfaggqyslkftdvsgtvridslwtrtkvsssvfstdpvavyqvnrvllpeaifgtdvoompAOAOAOivsAOSdSOSvadsegasspksshknsgqklllapismvisglvalfl"                                                                                                                  </span>
<span class="co">#&gt; [6] "mmrgaaptgvvsvmvlmilvllkqiesasangsslglpprkfcniyqgswvydksyplydskncpfierqfncksngrpdseylkyrwqpsgcnlprfngldflgrimkgkklmfvgdslslnqwqsltcllhnaapkanststrsosglsvfsfpaynssimfsrnaflvdivgappkrvmkldsissgslwktadvlvfnswhwwlhtdrkqpwdaimsgnvtvkdmdrlvayekammtwakwidqnidpsktkvffqgispdhgrarewskqggkgscigetkpimgssylagphaaemvvakviktmknqarlmdvtlmsqlrkdghpsvygfgghrmadcshwclsgvpdswnqllyselfhs"</span>
<span class="co">#&gt; [7] "maliknnifftsllifvtlfgvavggtvhkvgntkgwtmiggdyeawassrvfqvgdtlvfaynkdyhdvtevthndfemcesskplrryktgsdsisltkpglqhficgvpghckkgqklqihvlpaslghvavovogovrsqssssSOSOSOlvdpovnnapqyqmgotoashsaasadfiftfsfdltlidlctffilffilv"                                                                                                                                                                  </span>
<span class="co">#&gt; [8] "marsfaiavicivliagvtgqAOTSOOTaTOAOOTOTTOoOAaTOoovsAOoovttSOoovttAOoOAnoooovsSOoOASOoOATOoovaSOooovaSOoOATOoovaTOoOAOlaSOOAqvOAOAOTtkodSOSOSOSsSOolOSsdAOgOStdsiSOAOSOTdvndqngaskmvsslvfgsvlvwfmi"</span></code></pre></div>
<p>Extensin motifs in the form of SPPP+ are also detected by <code><a href="reference/scan_ag.html">scan_ag()</a></code> if in correct context, to avoid this add <code>exclude_ext = "yes"</code> as an argument:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">at_nsp_ag &lt;-<span class="st"> </span><span class="kw"><a href="reference/scan_ag.html">scan_ag</a></span>(hyp_pred<span class="op">$</span>sequence,
                     sequence, 
                     id,
                     <span class="dt">dim =</span> <span class="dv">3</span>, <span class="co">#at least 3 dimers must be present</span>
                     <span class="dt">div =</span> <span class="dv">10</span>, <span class="co">#no more than 10 amino acids apart</span>
                     <span class="dt">type =</span> <span class="st">"conservative"</span>,
                     <span class="dt">exclude_ext =</span> <span class="st">"yes"</span>)
<span class="co">#&gt; sequence vector contains O, O will be considered instead of P</span>

at_nsp_ag<span class="op">$</span>sequence[ind]
<span class="co">#&gt; [1] "mayatilmifsvvalmsgerahaavdcsslilnmadclsfvtsgstvvkpegtccsglktvvrtgpeclceafknsgslgltldlskaaslpsvckvaappsarcglsvsgdpOAtAOglSOTagagAOAlssganaatpvssprssdasllsvsfafvifmalissfy"                                                                                                                                                                                                       </span>
<span class="co">#&gt; [2] "mgyrrsyaitfvalvaalwsvtkaqpssscvstlttlspclsyitgnsttpsqpccsrldsviksspqcicsavnspipniglninrtqalqlpnacniqtppltqcnaatgOTaqpOAOSOTekTOdvtlTOTslogarsgvgggsktvpsvgtgsssrnvdplplhflmfavlvvctssfl"                                                                                                                                                                                         </span>
<span class="co">#&gt; [3] "mrlllsllfllalttyssatyclcrdgvgekdlqtsidyacgvlkdcnpihekgpcyqpntikshcdwavntyfqrfgqisgscnfsgtattsqnlpstvvtgclypsSOgsagtTOTtgTOSgtqtfpgoOAfgOAgdfdpsgnngapslfisialslgfsvviafl"                                                                                                                                                                                                        </span>
<span class="co">#&gt; [4] "mhhhhhpcnrkpfttifsffllylnlhnqqiiearnpsqfttnpsodvsipeikrhlqqygylpqnkesddvsfeqalvryqknlglpitgkpdsdtlsqillprcgfpddvepktapfhtgkkyvyfpgrprwtrdvplkltyafsqenltpylaptdirrvfrrafgkwasvipvsfietedyviadikigffngdhgdgepfdgvlgvlahtfspengrlhldkaetwavdfdeekssvavdlesvavheighvlglghssvkdaamyptlkprskkvnlnmddvvgvqslygtnpnftlnsllasetstnladgsrirsqgmiystlstvialcflnw"                          </span>
<span class="co">#&gt; [5] "makmqlsifiavvalivcsasaktaSOOAOvloOTOAOAOAOenvnltellsvagpfhtfldyllstgvietfqnqannteegitifvpkddafkaqknpplsnltkdqlkqlvlfhalphyyslsefknlsqsgpvstfaggqyslkftdvsgtvridslwtrtkvsssvfstdpvavyqvnrvllpeaifgtdvoompAOAOAOivsAOSdSOSvadsegasspksshknsgqklllapismvisglvalfl"                                                                                                                  </span>
<span class="co">#&gt; [6] "mmrgaaptgvvsvmvlmilvllkqiesasangsslglpprkfcniyqgswvydksyplydskncpfierqfncksngrpdseylkyrwqpsgcnlprfngldflgrimkgkklmfvgdslslnqwqsltcllhnaapkanststrsosglsvfsfpaynssimfsrnaflvdivgappkrvmkldsissgslwktadvlvfnswhwwlhtdrkqpwdaimsgnvtvkdmdrlvayekammtwakwidqnidpsktkvffqgispdhgrarewskqggkgscigetkpimgssylagphaaemvvakviktmknqarlmdvtlmsqlrkdghpsvygfgghrmadcshwclsgvpdswnqllyselfhs"</span>
<span class="co">#&gt; [7] "maliknnifftsllifvtlfgvavggtvhkvgntkgwtmiggdyeawassrvfqvgdtlvfaynkdyhdvtevthndfemcesskplrryktgsdsisltkpglqhficgvpghckkgqklqihvlpaslghvavovogovrsqssssSOSOSOlvdpovnnapqyqmgotoashsaasadfiftfsfdltlidlctffilffilv"                                                                                                                                                                  </span>
<span class="co">#&gt; [8] "marsfaiavicivliagvtgqAOTSOOTaTOAOOTOTTOoOAaTOoovsAOoovttsooovttaoooanoooovssoooasoooatooovasoooovasoooaTOoovaTOoOAOlaSOOAqvOAOAOTtkodSOSOSOSsSOolOSsdAOgOStdsiSOAOSOTdvndqngaskmvsslvfgsvlvwfmi"</span></code></pre></div>
<p>to switch of all PPP+ from being detected:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">at_nsp_ag &lt;-<span class="st"> </span><span class="kw"><a href="reference/scan_ag.html">scan_ag</a></span>(hyp_pred<span class="op">$</span>sequence,
                     sequence, 
                     id,
                     <span class="dt">dim =</span> <span class="dv">3</span>, <span class="co">#at least 3 dimers must be present</span>
                     <span class="dt">div =</span> <span class="dv">10</span>, <span class="co">#no more than 10 amino acids apart</span>
                     <span class="dt">type =</span> <span class="st">"conservative"</span>,
                     <span class="dt">exclude_ext =</span> <span class="st">"all"</span>)
<span class="co">#&gt; sequence vector contains O, O will be considered instead of P</span>

at_nsp_ag<span class="op">$</span>sequence[ind]
<span class="co">#&gt; [1] "mayatilmifsvvalmsgerahaavdcsslilnmadclsfvtsgstvvkpegtccsglktvvrtgpeclceafknsgslgltldlskaaslpsvckvaappsarcglsvsgdpOAtAOglSOTagagAOAlssganaatpvssprssdasllsvsfafvifmalissfy"                                                                                                                                                                                                       </span>
<span class="co">#&gt; [2] "mgyrrsyaitfvalvaalwsvtkaqpssscvstlttlspclsyitgnsttpsqpccsrldsviksspqcicsavnspipniglninrtqalqlpnacniqtppltqcnaatgOTaqpOAOSOTekTOdvtlTOTslogarsgvgggsktvpsvgtgsssrnvdplplhflmfavlvvctssfl"                                                                                                                                                                                         </span>
<span class="co">#&gt; [3] "mrlllsllfllalttyssatyclcrdgvgekdlqtsidyacgvlkdcnpihekgpcyqpntikshcdwavntyfqrfgqisgscnfsgtattsqnlpstvvtgclypsSOgsagtTOTtgTOSgtqtfpgoOAfgOAgdfdpsgnngapslfisialslgfsvviafl"                                                                                                                                                                                                        </span>
<span class="co">#&gt; [4] "mhhhhhpcnrkpfttifsffllylnlhnqqiiearnpsqfttnpsodvsipeikrhlqqygylpqnkesddvsfeqalvryqknlglpitgkpdsdtlsqillprcgfpddvepktapfhtgkkyvyfpgrprwtrdvplkltyafsqenltpylaptdirrvfrrafgkwasvipvsfietedyviadikigffngdhgdgepfdgvlgvlahtfspengrlhldkaetwavdfdeekssvavdlesvavheighvlglghssvkdaamyptlkprskkvnlnmddvvgvqslygtnpnftlnsllasetstnladgsrirsqgmiystlstvialcflnw"                          </span>
<span class="co">#&gt; [5] "makmqlsifiavvalivcsasaktaSOOAOvloOTOAOAOAOenvnltellsvagpfhtfldyllstgvietfqnqannteegitifvpkddafkaqknpplsnltkdqlkqlvlfhalphyyslsefknlsqsgpvstfaggqyslkftdvsgtvridslwtrtkvsssvfstdpvavyqvnrvllpeaifgtdvoompAOAOAOivsAOSdSOSvadsegasspksshknsgqklllapismvisglvalfl"                                                                                                                  </span>
<span class="co">#&gt; [6] "mmrgaaptgvvsvmvlmilvllkqiesasangsslglpprkfcniyqgswvydksyplydskncpfierqfncksngrpdseylkyrwqpsgcnlprfngldflgrimkgkklmfvgdslslnqwqsltcllhnaapkanststrsosglsvfsfpaynssimfsrnaflvdivgappkrvmkldsissgslwktadvlvfnswhwwlhtdrkqpwdaimsgnvtvkdmdrlvayekammtwakwidqnidpsktkvffqgispdhgrarewskqggkgscigetkpimgssylagphaaemvvakviktmknqarlmdvtlmsqlrkdghpsvygfgghrmadcshwclsgvpdswnqllyselfhs"</span>
<span class="co">#&gt; [7] "maliknnifftsllifvtlfgvavggtvhkvgntkgwtmiggdyeawassrvfqvgdtlvfaynkdyhdvtevthndfemcesskplrryktgsdsisltkpglqhficgvpghckkgqklqihvlpaslghvavovogovrsqssssSOSOSOlvdpovnnapqyqmgotoashsaasadfiftfsfdltlidlctffilffilv"                                                                                                                                                                  </span>
<span class="co">#&gt; [8] "marsfaiavicivliagvtgqAOTSOOTaTOAOOTOTtoooaatooovsaooovttsooovttaoooanoooovssoooasoooatooovasoooovasoooatooovatoooAOlaSOOAqvOAOAOTtkodSOSOSOSsSOolOSsdAOgOStdsiSOAOSOTdvndqngaskmvsslvfgsvlvwfmi"</span></code></pre></div>
</div>
<div id="scan-motifs-for-n-glycosylation" class="section level2">
<h2 class="hasAnchor">
<a href="#scan-motifs-for-n-glycosylation" class="anchor"></a>scan motifs for N-glycosylation</h2>
<p>N-glycosylation is frequent in secreted proteins. To scan the amino acid motifs for N-glycosylation the function <code><a href="reference/scan_nglc.html">scan_nglc()</a></code> can be used. Detection is based on <a href="https://prosite.expasy.org/PDOC00001">PROSITE pattern PS00001</a>. Mean local hydrophilicity <span class="citation">(Hopp and Woods 1981)</span> is used to assess if the asparagines are buried:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">at_nglc &lt;-<span class="st"> </span><span class="kw"><a href="reference/scan_nglc.html">scan_nglc</a></span>(at_nsp,
                     sequence, 
                     Transcript.id)</code></pre></div>
</div>
<div id="create-a-protein-structure-diagram" class="section level2">
<h2 class="hasAnchor">
<a href="#create-a-protein-structure-diagram" class="anchor"></a>create a protein structure diagram</h2>
<p>Using the above mentioned functions several protein features can be obtained. To visualize these features the function <code><a href="reference/plot_prot.html">plot_prot()</a></code> can be used. Domains (<code><a href="reference/get_hmm.html">get_hmm()</a></code>), N-terminal signal peptides (<code><a href="reference/get_signalp.html">get_signalp()</a></code>), transmembrane regions, extracellular and intracellular protein regions (<code><a href="reference/get_phobius.html">get_phobius()</a></code>), GPI attachment sites (<code><a href="reference/get_big_pi.html">get_big_pi()</a></code> or <code><a href="reference/get_pred_gpi.html">get_pred_gpi()</a></code>), AG glycomodul spans (<code><a href="reference/scan_ag.html">scan_ag()</a></code>), hydroxyproline positions (<code><a href="reference/predict_hyp.html">predict_hyp()</a></code>) and disordered reagions (<code><a href="reference/get_espritz.html">get_espritz()</a></code>, off by default) can be shown. The output is a <code>ggplot2</code> object which can be manipulated to customize the theme, colors and plot annotations.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ind &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dv">23</span>, <span class="dv">24</span>, <span class="dv">5</span>, <span class="dv">80</span>, <span class="dv">81</span>, <span class="dv">120</span>, <span class="dv">230</span>, <span class="dv">334</span>, <span class="dv">345</span>, <span class="dv">1000</span>)
pred &lt;-<span class="st"> </span><span class="kw"><a href="reference/plot_prot.html">plot_prot</a></span>(<span class="dt">sequence =</span> at_nsp<span class="op">$</span>sequence[ind],
                  <span class="dt">id =</span> at_nsp<span class="op">$</span>Transcript.id[ind])

pred</code></pre></div>
<div class="figure">
<img src="https://missuse.github.io/ragp/reference/figures/README-plot_prot-1.png" alt="alt text"><p class="caption">alt text</p>
</div>
</div>
<div id="acknowledgements" class="section level2">
<h2 class="hasAnchor">
<a href="#acknowledgements" class="anchor"></a>Acknowledgements</h2>
<p>This software was developed with funding from the Ministry of Education, Science and Technological Development of the Republic of Serbia (Projects TR31019 and OI173024).</p>
</div>
<div id="references" class="section level2 unnumbered">
<h2 class="hasAnchor">
<a href="#references" class="anchor"></a>References:</h2>
<div id="refs" class="references">
<div id="ref-eisenhaber_glycosylphosphatidylinositol_2003">
<p>Eisenhaber, Birgit, Michael Wildpaner, Carolyn J. Schultz, Georg H. H. Borner, Paul Dupree, and Frank Eisenhaber. 2003. “Glycosylphosphatidylinositol Lipid Anchoring of Plant Proteins. Sensitive Prediction from Sequence- and Genome-Wide Studies for Arabidopsis and Rice.” <em>Plant Physiology</em> 133 (4): 1691–1701. doi:<a href="https://doi.org/10.1104/pp.103.023580">10.1104/pp.103.023580</a>.</p>
</div>
<div id="ref-ellis_arabinogalactan-proteins:_2010">
<p>Ellis, Miriam, Jack Egelund, Carolyn J. Schultz, and Antony Bacic. 2010. “Arabinogalactan-Proteins: Key Regulators at the Cell Surface?” <em>Plant Physiology</em> 153 (2): 403–19. doi:<a href="https://doi.org/10.1104/pp.110.156000">10.1104/pp.110.156000</a>.</p>
</div>
<div id="ref-emanuelsson_predicting_2000">
<p>Emanuelsson, O., H. Nielsen, S. Brunak, and G. von Heijne. 2000. “Predicting Subcellular Localization of Proteins Based on Their N-Terminal Amino Acid Sequence.” <em>Journal of Molecular Biology</em> 300 (4): 1005–16. doi:<a href="https://doi.org/10.1006/jmbi.2000.3903">10.1006/jmbi.2000.3903</a>.</p>
</div>
<div id="ref-finn_hmmer_2011">
<p>Finn, Robert D., Jody Clements, and Sean R. Eddy. 2011. “HMMER Web Server: Interactive Sequence Similarity Searching.” <em>Nucleic Acids Research</em> 39 (Web Server issue): W29–W37. doi:<a href="https://doi.org/10.1093/nar/gkr367">10.1093/nar/gkr367</a>.</p>
</div>
<div id="ref-hopp_prediction_1981">
<p>Hopp, T. P., and K. R. Woods. 1981. “Prediction of Protein Antigenic Determinants from Amino Acid Sequences.” <em>Proceedings of the National Academy of Sciences of the United States of America</em> 78 (6): 3824–8.</p>
</div>
<div id="ref-johnson_pipeline_2017">
<p>Johnson, Kim L., Andrew M. Cassin, Andrew Lonsdale, Antony Bacic, Monika S. Doblin, and Carolyn J. Schultz. 2017. “Pipeline to Identify Hydroxyproline-Rich Glycoproteins.” <em>Plant Physiology</em> 174 (2): 886–903. doi:<a href="https://doi.org/10.1104/pp.17.00294">10.1104/pp.17.00294</a>.</p>
</div>
<div id="ref-kall_advantages_2007">
<p>Käll, Lukas, Anders Krogh, and Erik L. L. Sonnhammer. 2007. “Advantages of Combined Transmembrane Topology and Signal Peptide Prediction–the Phobius Web Server.” <em>Nucleic Acids Research</em> 35 (Web Server issue): W429–432. doi:<a href="https://doi.org/10.1093/nar/gkm256">10.1093/nar/gkm256</a>.</p>
</div>
<div id="ref-pierleoni_predgpi_2008">
<p>Pierleoni, Andrea, Pier Luigi Martelli, and Rita Casadio. 2008. “PredGPI: A GPI-Anchor Predictor.” <em>BMC Bioinformatics</em> 9 (September): 392. doi:<a href="https://doi.org/10.1186/1471-2105-9-392">10.1186/1471-2105-9-392</a>.</p>
</div>
<div id="ref-walsh_espritz_2012">
<p>Walsh, Ian, Alberto J. M. Martin, Tomàs Di Domenico, and Silvio C. E. Tosatto. 2012. “ESpritz: Accurate and Fast Prediction of Protein Disorder.” <em>Bioinformatics</em> 28 (4): 503–9. doi:<a href="https://doi.org/10.1093/bioinformatics/btr682">10.1093/bioinformatics/btr682</a>.</p>
</div>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Browse source code at <br><a href="https://github.com/missuse/ragp">https://​github.com/​missuse/​ragp</a>
</li>
<li>Report a bug at <br><a href="https://github.com/missuse/ragp/issues">https://​github.com/​missuse/​ragp/​issues</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li>file <a href="LICENSE-text.html">LICENSE</a>
</li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Milan Dragicevic <br><small class="roles"> Author, maintainer </small>  </li>
<li>Danijela Paunovic <br><small class="roles"> Author </small>  </li>
<li><a href="authors.html">All authors...</a></li>
</ul>
</div>

      <div class="dev-status">
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://travis-ci.org/missuse/ragp"><img src="https://travis-ci.org/missuse/ragp.svg?branch=master" alt="Build Status"></a></li>
<li><a href="http://www.repostatus.org/#active"><img src="http://www.repostatus.org/badges/latest/active.svg" alt="Project Status: Active – The project has reached a stable, usable state and is being actively developed."></a></li>
<li><a href="https://opensource.org/licenses/MIT"><img src="https://img.shields.io/badge/License-MIT-blue.svg" alt="License: MIT"></a></li>
</ul>
</div>
</div>

</div>


      <footer><div class="copyright">
  <p>Developed by Milan Dragicevic, Danijela Paunovic.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  

  </body>
</html>
