<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Query hmmer web server. — get_hmm • ragp</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Query hmmer web server. — get_hmm" />

<meta property="og:description" content="hmmer web server offers biosequence analysis using profile hidden Markov Models. This function allows searching
of a protein sequence vs a profile-HMM database (Pfam-A)." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-inverse navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ragp</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.2.1.9100</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/pkgdown/start.html">Getting started</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/pkgdown/filter.html">Filtering hydroxyproline rich glycoproteins</a>
    </li>
    <li>
      <a href="../articles/pkgdown/analyse.html">Analyse hydroxyproline rich glycoproteins</a>
    </li>
    <li>
      <a href="../articles/pkgdown/PSV.html">Protein Sequence Visualization</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
<li>
  <a href="https://ragp.shinyapps.io/Rapp/">ragp shiny</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/missuse/ragp/">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Query hmmer web server.</h1>
    <small class="dont-index">Source: <a href='https://github.com/missuse/ragp/blob/master/R/get_hmm.R'><code>R/get_hmm.R</code></a></small>
    <div class="hidden name"><code>get_hmm.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>hmmer web server offers biosequence analysis using profile hidden Markov Models. This function allows searching
of a protein sequence vs a profile-HMM database (Pfam-A).</p>
    
    </div>

    <pre class="usage"><span class='fu'>get_hmm</span>(<span class='no'>data</span>, <span class='no'>...</span>)

<span class='co'># S3 method for character</span>
<span class='fu'>get_hmm</span>(<span class='no'>data</span>, <span class='no'>...</span>)

<span class='co'># S3 method for data.frame</span>
<span class='fu'>get_hmm</span>(<span class='no'>data</span>, <span class='no'>sequence</span>, <span class='no'>id</span>, <span class='no'>...</span>)

<span class='co'># S3 method for list</span>
<span class='fu'>get_hmm</span>(<span class='no'>data</span>, <span class='no'>...</span>)

<span class='co'># S3 method for default</span>
<span class='fu'>get_hmm</span>(<span class='kw'>data</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='no'>sequence</span>, <span class='no'>id</span>, <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>sleep</span> <span class='kw'>=</span> <span class='fl'>1</span>, <span class='kw'>attempts</span> <span class='kw'>=</span> <span class='fl'>2L</span>, <span class='kw'>timeout</span> <span class='kw'>=</span> <span class='fl'>10</span>, <span class='kw'>progress</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='no'>...</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>data</th>
      <td><p>A data frame with protein amino acid sequences as strings in one column and corresponding id's in another. Alternatively a path to a .fasta file with protein sequences. Alternatively a list with elements of class "SeqFastaAA" resulting from <code><a href='http://www.rdocumentation.org/packages/seqinr/topics/read.fasta'>read.fasta</a></code> call. Should be left blank if vectors are provided to sequence and id arguments.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>currently no additional arguments are accepted apart the ones documented bellow.</p></td>
    </tr>
    <tr>
      <th>sequence</th>
      <td><p>A vector of strings representing protein amino acid sequences, or the appropriate column name if a data.frame is supplied to data argument. If .fasta file path, or list with elements of class "SeqFastaAA" provided to data, this should be left blank.</p></td>
    </tr>
    <tr>
      <th>id</th>
      <td><p>A vector of strings representing protein identifiers, or the appropriate column name if a data.frame is supplied to data argument. If .fasta file path, or list with elements of class "SeqFastaAA" provided to data, this should be left blank.</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>Bolean, whether to print out the output for each sequence, defaults to FALSE.</p></td>
    </tr>
    <tr>
      <th>sleep</th>
      <td><p>Numeric indicating the pause in seconds between server calls, at default set to 1.</p></td>
    </tr>
    <tr>
      <th>attempts</th>
      <td><p>Integer, number of attempts if server unresponsive, at default set to 2.</p></td>
    </tr>
    <tr>
      <th>timeout</th>
      <td><p>Numeric, time in seconds to wait for server response.</p></td>
    </tr>
    <tr>
      <th>progress</th>
      <td><p>Bolean, whether to show the progress bar, at default set to FALSE.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="source"><a class="anchor" href="#source"></a>Source</h2>

    <p><a href='https://www.ebi.ac.uk/Tools/hmmer/search/hmmscan'>https://www.ebi.ac.uk/Tools/hmmer/search/hmmscan</a></p>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A data frame with columns:</p><dl class='dl-horizontal'>
  <dt>id</dt><dd><p>Character, as supplied in the function call</p></dd>
  <dt>name</dt><dd><p>Character, PFAM family name</p></dd>
  <dt>acc</dt><dd><p>Character, PFAM family accession</p></dd>
  <dt>desc</dt><dd><p>Character, PFAM family description</p></dd>
  <dt>clan</dt><dd><p>Character, PFAM clan</p></dd>
  <dt>align_start</dt><dd><p>Numeric, start of domain alignment in query sequence</p></dd>
  <dt>align_end</dt><dd><p>Numeric, end of domain alignment in query sequence</p></dd>
  <dt>model_start</dt><dd><p>Numeric, start of alignment in domain model</p></dd>
  <dt>model_end</dt><dd><p>Numeric, end of alignment in domain model</p></dd>
  <dt>ievalue</dt><dd><p>Numeric, the "independent E-value", the E-value that the sequence/profile comparison would have received if this were the only domain envelope found in it, excluding any others. This is a stringent measure of how reliable this particular domain may be. The independent E-value uses the total number of targets in the target database.</p></dd>
  <dt>cevalue</dt><dd><p>Numeric, the "conditional E-value", a permissive measure of how reliable this particular domain may be.</p></dd>
  <dt>bitscore</dt><dd><p>Numeric, the domain bit score.</p></dd>
  <dt>reported</dt><dd><p>Logical, is the result reported on the hmmer site. The hmmer web server outputs more hmm profile matches than it presents to the user. Results below a certain threshold are not reported (hidden) on the site.</p></dd>
</dl>

    
    <h2 class="hasAnchor" id="note"><a class="anchor" href="#note"></a>Note</h2>

    <p>hmmscan does not handle sequences longer than 1000 amino acids. get_hmm splits these sequences into shorter substrings which overlap by 300 amino acids and queries hmmscan. Some results might be redundant or partially overlapping in this case. When this is an issue it is advisable to provide a subsequence of appropriate length as get_hmm input.</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='pfam2go.html'>pfam2go</a></code></p></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'>
<span class='no'>pfam_pred</span> <span class='kw'>&lt;-</span> <span class='fu'>get_hmm</span>(<span class='kw'>data</span> <span class='kw'>=</span> <span class='no'>at_nsp</span>[<span class='fl'>1</span>:<span class='fl'>5</span>,],
                    <span class='kw'>sequence</span> <span class='kw'>=</span> <span class='no'>sequence</span>,
                    <span class='kw'>id</span> <span class='kw'>=</span> <span class='no'>Transcript.id</span>,
                    <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
                    <span class='kw'>sleep</span> <span class='kw'>=</span> <span class='fl'>0</span>)
<span class='no'>pfam_pred</span></div><div class='output co'>#&gt;            id           name        acc                    desc   clan
#&gt; 1 ATCG00660.1  Ribosomal_L20 PF00453.18   Ribosomal protein L20   &lt;NA&gt;
#&gt; 2 AT2G43600.1 Glyco_hydro_19 PF00182.19       Chitinase class I   &lt;NA&gt;
#&gt; 3 AT2G43600.1 Glyco_hydro_19 PF00182.19       Chitinase class I CL0037
#&gt; 4 AT2G43600.1 Glyco_hydro_19 PF00182.19       Chitinase class I CL0037
#&gt; 5 AT2G28410.1           &lt;NA&gt;       &lt;NA&gt;                    &lt;NA&gt;   &lt;NA&gt;
#&gt; 6 AT2G22960.1  Peptidase_S10 PF00450.22 Serine carboxypeptidase   &lt;NA&gt;
#&gt; 7 AT2G22960.1  Peptidase_S10 PF00450.22 Serine carboxypeptidase CL0028
#&gt; 8 AT2G19580.1    Tetraspanin PF00335.20      Tetraspanin family CL0347
#&gt;   align_start align_end model_start model_end ievalue cevalue    bitscore
#&gt; 1           3       108           1       104 3.2e-32 1.8e-36 110.7813873
#&gt; 2          37        48          75        86 9.1e+03 5.1e-01  -3.7373791
#&gt; 3          83       220           1       157 2.4e-29 1.3e-33 102.9026337
#&gt; 4         227       273         185       232 7.9e-05 4.4e-09  22.6424046
#&gt; 5          NA        NA          NA        NA      NA      NA          NA
#&gt; 6          30        48           1        19 4.8e+02 2.7e-02  -0.1215759
#&gt; 7          48       181         279       419 4.8e-37 2.7e-41 128.3546143
#&gt; 8           6       251           3       228 4.5e-30 2.5e-34 105.0756226
#&gt;   reported
#&gt; 1     TRUE
#&gt; 2    FALSE
#&gt; 3     TRUE
#&gt; 4     TRUE
#&gt; 5    FALSE
#&gt; 6    FALSE
#&gt; 7     TRUE
#&gt; 8     TRUE</div><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#source">Source</a></li>

      <li><a href="#value">Value</a></li>

      <li><a href="#note">Note</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Milan Dragicevic, Danijela Paunovic.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

